From b458f6a67926a1cc1cc33f29f78175feb21ab639 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Fri, 4 Mar 2016 05:57:50 -0500
Subject: [PATCH] Absorption API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index 5410e93b..f910eca0 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -237,6 +237,16 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
     }
 
     @Override
+    public float getAbsorption() {
+        return getHandle().getAbsorptionHearts();
+    }
+
+    @Override
+    public void setAbsorption(float absorption) {
+        getHandle().setAbsorptionHearts(absorption);
+    }
+
+    @Override
     public EntityLiving getHandle() {
         return (EntityLiving) entity;
     }
