From 1d01d91c035d52edd2d3a7f28e98bfdfd6866c42 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Fri, 4 Mar 2016 05:57:50 -0500
Subject: [PATCH] Absorption API


diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
index e6ca2a9e..ed2ad306 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftLivingEntity.java
@@ -232,6 +232,16 @@ public class CraftLivingEntity extends CraftEntity implements LivingEntity {
         getHandle().noDamageTicks = ticks;
     }
 
+    @Override
+    public float getAbsorption() {
+        return getHandle().getAbsorptionHearts();
+    }
+
+    @Override
+    public void setAbsorption(float absorption) {
+        getHandle().setAbsorptionHearts(absorption);
+    }
+
     @Override
     public EntityLiving getHandle() {
         return (EntityLiving) entity;
