From 6c12b4df6f813f306e1190bf3f981aaf28c14edf Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Wed, 4 Feb 2015 03:39:46 -0500
Subject: [PATCH] Add setting to disable skull fetching


diff --git a/src/main/java/net/minecraft/server/TileEntitySkull.java b/src/main/java/net/minecraft/server/TileEntitySkull.java
index 223e2ed2..c87d9461 100644
--- a/src/main/java/net/minecraft/server/TileEntitySkull.java
+++ b/src/main/java/net/minecraft/server/TileEntitySkull.java
@@ -106,6 +106,9 @@ public class TileEntitySkull extends TileEntity implements ITickable {
         if (gameprofile != null && !UtilColor.b(gameprofile.getName())) {
             if (gameprofile.isComplete() && gameprofile.getProperties().containsKey("textures")) {
                 return gameprofile;
+            } else if(!org.bukkit.Bukkit.getFetchSkulls()) {
+                org.bukkit.Bukkit.getLogger().warning("Skull-fetching is disabled, so not fetching skin for '" + gameprofile.getName() + "'");
+                return gameprofile;
             } else if (TileEntitySkull.j != null && TileEntitySkull.k != null) {
                 GameProfile gameprofile1 = TileEntitySkull.j.getProfile(gameprofile.getName());
 
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 0003602b..81b11739 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -587,6 +587,11 @@ public final class CraftServer implements Server {
     }
 
     @Override
+    public boolean getFetchSkulls() {
+        return this.configuration.getBoolean("settings.fetch-skulls", true);
+    }
+
+    @Override
     public int getTicksPerAnimalSpawns() {
         return this.configuration.getInt("ticks-per.animal-spawns");
     }
diff --git a/src/main/resources/configurations/bukkit.yml b/src/main/resources/configurations/bukkit.yml
index 5230fd4d..051c3977 100644
--- a/src/main/resources/configurations/bukkit.yml
+++ b/src/main/resources/configurations/bukkit.yml
@@ -24,6 +24,7 @@ settings:
     deprecated-verbose: default
     shutdown-message: Server closed
     bungeecord: false
+    fetch-skulls: true
 spawn-limits:
     monsters: 70
     animals: 15
