From 5384f83ca291cef817ba6bc1ff853d9a967ff217 Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Wed, 31 Jan 2018 23:42:29 -0800
Subject: [PATCH] Add setting to disable advancements


diff --git a/src/main/java/net/minecraft/server/AdvancementDataPlayer.java b/src/main/java/net/minecraft/server/AdvancementDataPlayer.java
index 495e7f8..38b2416 100644
--- a/src/main/java/net/minecraft/server/AdvancementDataPlayer.java
+++ b/src/main/java/net/minecraft/server/AdvancementDataPlayer.java
@@ -165,6 +165,7 @@ public class AdvancementDataPlayer {
     }
 
     public void c() {
+        if (!org.bukkit.Bukkit.getAdvancementsEnabled()) return;
         HashMap hashmap = Maps.newHashMap();
         Iterator iterator = this.data.entrySet().iterator();
 
diff --git a/src/main/java/net/minecraft/server/AdvancementDataWorld.java b/src/main/java/net/minecraft/server/AdvancementDataWorld.java
index 5ae0609..940a463 100644
--- a/src/main/java/net/minecraft/server/AdvancementDataWorld.java
+++ b/src/main/java/net/minecraft/server/AdvancementDataWorld.java
@@ -150,6 +150,9 @@ public class AdvancementDataWorld {
                     java.nio.file.Path java_nio_file_path1 = (java.nio.file.Path) iterator.next();
 
                     if ("json".equals(FilenameUtils.getExtension(java_nio_file_path1.toString()))) {
+                        if (!org.bukkit.Bukkit.getAdvancementsEnabled()) {
+                            continue;
+                        }
                         java.nio.file.Path java_nio_file_path2 = java_nio_file_path.relativize(java_nio_file_path1);
                         String s = FilenameUtils.removeExtension(java_nio_file_path2.toString()).replaceAll("\\\\", "/");
                         MinecraftKey minecraftkey = new MinecraftKey("minecraft", s);
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftServer.java b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
index 69c880e..c63fa44 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftServer.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftServer.java
@@ -644,6 +644,11 @@ public final class CraftServer extends CraftBukkitRuntime implements Server {
     }
 
     @Override
+    public boolean getAdvancementsEnabled() {
+        return this.configuration.getBoolean("settings.advancements-enabled", true);
+    }
+
+    @Override
     public int getTicksPerAnimalSpawns() {
         return this.configuration.getInt("ticks-per.animal-spawns");
     }
