From: Jedediah Smith <jedediah@silencegreys.com>
Date: Wed, 22 Jun 2016 15:17:26 -0400
Subject: [PATCH] Allow empty ItemStack


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -155,9 +155,6 @@ public final class CraftItemStack extends ItemStack {
         }
 
         handle.setCount(amount);
-        if (amount == 0) {
-            handle = null;
-        }
     }
 
     @Override
