From 13532af4080d0d6788745e2d7958d223e03129df Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Wed, 22 Jun 2016 15:17:26 -0400
Subject: [PATCH] Allow empty ItemStack


diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
index 252ed1d9..f22d51e1 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftItemStack.java
@@ -155,9 +155,6 @@ public final class CraftItemStack extends ItemStack {
         }
 
         handle.setCount(amount);
-        if (amount == 0) {
-            handle = null;
-        }
     }
 
     @Override
