From ade5e8dfbeeb69230038a05a90fdcac14545f151 Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Thu, 1 Feb 2018 10:50:40 -0800
Subject: [PATCH] Catch Async PlayerChunkMap operations


diff --git a/src/main/java/net/minecraft/server/PlayerChunkMap.java b/src/main/java/net/minecraft/server/PlayerChunkMap.java
index 35b15a7..558bfde 100644
--- a/src/main/java/net/minecraft/server/PlayerChunkMap.java
+++ b/src/main/java/net/minecraft/server/PlayerChunkMap.java
@@ -421,10 +421,12 @@ public class PlayerChunkMap {
     }
 
     public void a(PlayerChunk playerchunk) {
+        org.bukkit.craftbukkit.AsyncCatcher.catchOp("Async Player Chunk Add"); // Paper
         this.f.add(playerchunk);
     }
 
     public void b(PlayerChunk playerchunk) {
+        org.bukkit.craftbukkit.AsyncCatcher.catchOp("Async Player Chunk Remove"); // Paper
         ChunkCoordIntPair chunkcoordintpair = playerchunk.a();
         long i = d(chunkcoordintpair.x, chunkcoordintpair.z);
 
