From d9ba108388cdf1e5708d5f014279dd1f699c48bb Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Wed, 30 Apr 2014 17:37:27 -0500
Subject: [PATCH] Disable resource leak detector


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 1f7c541f..bf3f88f1 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -132,6 +132,10 @@ public abstract class MinecraftServer implements ICommandListener, Runnable, IAs
         this.X = gameprofilerepository;
         this.Y = usercache;
         // this.universe = file; // CraftBukkit
+
+        // SportBukkit - disable netty's resource leak detection
+        io.netty.util.ResourceLeakDetector.setEnabled(false);
+
         this.p = new ServerConnection(this);
         this.b = this.i();
         // this.convertable = new WorldLoaderServer(file); // CraftBukkit - moved to DedicatedServer.init
