From d61d79df9bd735e853949d9b09f0f92531896024 Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Thu, 26 Jul 2018 00:37:50 -0700
Subject: [PATCH] Dont kick for tab complete spamming


diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
index 97b0ac76..a7f8ed67 100644
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -2326,7 +2326,7 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
         PlayerConnectionUtils.ensureMainThread(packetplayintabcomplete, this, this.player.x());
         // CraftBukkit start
         if (chatSpamField.addAndGet(this, 10) > 500 && !this.minecraftServer.getPlayerList().isOp(this.player.getProfile())) {
-            this.disconnect(new ChatMessage("disconnect.spam", new Object[0]));
+            //this.disconnect(new ChatMessage("disconnect.spam", new Object[0])); 1.13 spams tabcomplete packets
             return;
         }
         // CraftBukkit end
