From: Chad Waters <authorblues@gmail.com>
Date: Sun, 20 Jan 2013 22:04:58 -0500
Subject: [PATCH] Fire ServerCommandEvent for command block cmds. Addresses
 BUKKIT-3165


diff --git a/src/main/java/net/minecraft/server/CommandBlockListenerAbstract.java b/src/main/java/net/minecraft/server/CommandBlockListenerAbstract.java
--- a/src/main/java/net/minecraft/server/CommandBlockListenerAbstract.java
+++ b/src/main/java/net/minecraft/server/CommandBlockListenerAbstract.java
@@ -181,14 +181,6 @@ public abstract class CommandBlockListenerAbstract implements ICommandListener {
         if (command.startsWith("/")) {
             command = command.substring(1);
         }
-
-        ServerCommandEvent event = new ServerCommandEvent(bSender, command);
-        org.bukkit.Bukkit.getPluginManager().callEvent(event);
-        if (event.isCancelled()) {
-            return 0;
-        }
-        command = event.getCommand();
-
         String[] args = command.split(" ");
         ArrayList<String[]> commands = new ArrayList<String[]>();

@@ -264,6 +272,10 @@ public abstract class CommandBlockListenerAbstract implements ICommandListener {
         // Now dispatch all of the commands we ended up with
         for (int i = 0; i < commands.size(); i++) {
             try {
+                String commandLine = joiner.join(java.util.Arrays.asList(commands.get(i)));
+                org.bukkit.event.server.ServerCommandEvent event = new org.bukkit.event.server.ServerCommandEvent(bSender, commandLine);
+                bSender.getServer().getPluginManager().callEvent(event);
+
                 if (commandMap.dispatch(bSender, joiner.join(java.util.Arrays.asList(commands.get(i))))) {
                     completed++;
                 }
