From aa380d68183f1ea49679e91ef1200f728980ebcb Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 17 Feb 2015 03:50:47 -0500
Subject: [PATCH] Fix exception playing particle effects


diff --git a/src/main/java/org/bukkit/craftbukkit/CraftEffect.java b/src/main/java/org/bukkit/craftbukkit/CraftEffect.java
index 582d7bfb..845245cd 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftEffect.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftEffect.java
@@ -65,4 +65,13 @@ public class CraftEffect {
         }
         return datavalue;
     }
+
+    public static net.minecraft.server.EnumParticle getNMSParticle(Effect effect) {
+        String name = effect.getName();
+        if(name == null) return null;
+        for(net.minecraft.server.EnumParticle nms : net.minecraft.server.EnumParticle.values()) {
+            if(name.equals(nms.b())) return nms;
+        }
+        return null;
+    }
 }
diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index 11f17e23..85f9574b 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -876,7 +876,7 @@ public class CraftWorld implements World {
                 particleFullName.append('_').append(data);
             }
 
-            packet = new PacketPlayOutWorldParticles(EnumParticle.valueOf(effect.getName()), false, (float)location.getX(), (float)location.getY(), (float)location.getZ(), offsetX, offsetY, offsetZ, particleCount, radius);
+            packet = new PacketPlayOutWorldParticles(CraftEffect.getNMSParticle(effect), false, (float)location.getX(), (float)location.getY(), (float)location.getZ(), offsetX, offsetY, offsetZ, particleCount, radius);
         }
 
         int distance;
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
index 9d0d4cf0..e59c768f 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftPlayer.java
@@ -684,7 +684,7 @@ public class CraftPlayer extends CraftHumanEntity implements Player {
             if (effect.getData() != null && effect.getData().equals(MaterialData.class)) {
                 particleFullName.append('_').append(data);
             }
-            packet = new PacketPlayOutWorldParticles(EnumParticle.valueOf(effect.getName()), false, (float)location.getX(), (float)location.getY(), (float)location.getZ(), offsetX, offsetY, offsetZ, particleCount, radius);
+            packet = new PacketPlayOutWorldParticles(CraftEffect.getNMSParticle(effect), false, (float)location.getX(), (float)location.getY(), (float)location.getZ(), offsetX, offsetY, offsetZ, particleCount, radius);
         }
 
         if (!location.getWorld().equals(getWorld())) {
