From: Jedediah Smith <jedediah@silencegreys.com>
Date: Wed, 23 Nov 2016 04:38:06 -0500
Subject: [PATCH] Fix no knockback from zero-damage projectiles


diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -1418,7 +1418,7 @@ public abstract class EntityLiving extends Entity {
 
                     return false;
                 } else {
-                    return originalDamage > 0;
+                    return true;
                 }
                 // CraftBukkit end
             }
