From e663d9d7edf19ceabec7e0e3e6afab1bf394ef35 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Wed, 23 Nov 2016 04:38:06 -0500
Subject: [PATCH] Fix no knockback from zero-damage projectiles


diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index 93e0bafa..ece56fb8 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -1421,7 +1421,7 @@ public abstract class EntityLiving extends Entity {
 
                     return false;
                 } else {
-                    return originalDamage > 0;
+                    return true;
                 }
                 // CraftBukkit end
             }
