From 40ac371aea01c0d0f51c6efc6ec35a4372815034 Mon Sep 17 00:00:00 2001
From: ShinyDialga <shinydialga45@gmail.com>
Date: Wed, 25 Jul 2018 15:38:58 -0500
Subject: [PATCH] Fix sprint disabling when engaging in combat


diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index 8bc82c02..b6f76516 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -1039,7 +1039,7 @@ public abstract class EntityHuman extends EntityLiving {
 
                             this.motX *= 0.6D;
                             this.motZ *= 0.6D;
-                            this.setSprinting(false);
+//                            this.setSprinting(false); //SportBukkit - fix sprinting bug
                         }
 
                         if (flag3) {
