From 5d2cf0f8f772eabd49b57b631415a5c87de892ea Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Fri, 10 Jul 2015 23:32:00 -0400
Subject: [PATCH] Fix self-damage being cancelled by FF/PvP rules


diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
index 8ab0f81d..78b57910 100644
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -777,6 +777,9 @@ public abstract class EntityHuman extends EntityLiving {
     public boolean a(EntityHuman entityhuman) {
         // CraftBukkit start - Change to check OTHER player's scoreboard team according to API
         // To summarize this method's logic, it's "Can parameter hurt this"
+
+        if(this == entityhuman) return true; // SportBukkit - self-damage is always allowed
+
         org.bukkit.scoreboard.Team team;
         if (entityhuman instanceof EntityPlayer) {
             EntityPlayer thatPlayer = (EntityPlayer) entityhuman;
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index 5b55295a..9a6e50d9 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -583,6 +583,7 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
     }
 
     public boolean a(EntityHuman entityhuman) {
+        if(this == entityhuman) return true; // SportBukkit - self-damage is always allowed
         return !this.canPvP() ? false : super.a(entityhuman);
     }
 
