From ee8523a2fab10b893948fba4ad4aa8c4711479a3 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Fri, 9 May 2014 22:38:54 -0400
Subject: [PATCH] Fix wrong potion particles after respawn


diff --git a/src/main/java/net/minecraft/server/EntityLiving.java b/src/main/java/net/minecraft/server/EntityLiving.java
index ab7797b9..8767c367 100644
--- a/src/main/java/net/minecraft/server/EntityLiving.java
+++ b/src/main/java/net/minecraft/server/EntityLiving.java
@@ -628,6 +628,7 @@ public abstract class EntityLiving extends Entity {
         return false;
     }
 
+    public void clearPotionEffectMetadata() { this.bY(); } // SportBukkit - alias for following method
     protected void bY() {
         this.datawatcher.set(EntityLiving.h, Boolean.valueOf(false));
         this.datawatcher.set(EntityLiving.g, Integer.valueOf(0));
diff --git a/src/main/java/net/minecraft/server/EntityPlayer.java b/src/main/java/net/minecraft/server/EntityPlayer.java
index c3312439..5a6e9efa 100644
--- a/src/main/java/net/minecraft/server/EntityPlayer.java
+++ b/src/main/java/net/minecraft/server/EntityPlayer.java
@@ -1475,6 +1475,14 @@ public class EntityPlayer extends EntityHuman implements ICrafting {
         this.f(0); // PAIL: setArrowCount
         this.removeAllEffects();
         this.updateEffects = true;
+
+        // SportBukkit - clear potion metadata now, because new effects might
+        // get added before the update in the tick has a chance to run, and if they
+        // match the old effects, the metadata will never be marked dirty and will
+        // go out of sync with the client.
+        this.clearPotionEffectMetadata();
+        this.setInvisible(false);
+
         this.activeContainer = this.defaultContainer;
         this.killer = null;
         this.lastDamager = null;
