From dc903f165f039754a6f3403b724bf3dfdc11c36e Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Wed, 18 Feb 2015 06:36:47 -0500
Subject: [PATCH] Include stack traces in packet encoding exceptions


diff --git a/src/main/java/net/minecraft/server/PacketEncoder.java b/src/main/java/net/minecraft/server/PacketEncoder.java
index c81822d6..75e6ecaa 100644
--- a/src/main/java/net/minecraft/server/PacketEncoder.java
+++ b/src/main/java/net/minecraft/server/PacketEncoder.java
@@ -41,7 +41,7 @@ public class PacketEncoder extends MessageToByteEncoder<Packet<?>> {
                 try {
                     packet.b(packetdataserializer);
                 } catch (Throwable throwable) {
-                    PacketEncoder.a.error(throwable);
+                PacketEncoder.a.error("Exception serializing packet " + packet, throwable); // SportBukkit - better error message
                 }
 
             }
