From d0efc26d8ba2167294e13672e1a81ed13e8d013b Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sat, 7 Dec 2013 17:53:32 -0600
Subject: [PATCH] Make scoreboard classes public


diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftCriteria.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftCriteria.java
index 7dedd022..5a7c60c2 100644
--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftCriteria.java
+++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftCriteria.java
@@ -7,7 +7,7 @@ import net.minecraft.server.ScoreboardObjective;
 
 import com.google.common.collect.ImmutableMap;
 
-final class CraftCriteria {
+public final class CraftCriteria {
     static final Map<String, CraftCriteria> DEFAULTS;
     static final CraftCriteria DUMMY;
 
diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftObjective.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftObjective.java
index 239ceca0..27641d85 100644
--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftObjective.java
+++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftObjective.java
@@ -9,7 +9,7 @@ import org.bukkit.scoreboard.DisplaySlot;
 import org.bukkit.scoreboard.Objective;
 import org.bukkit.scoreboard.Score;
 
-final class CraftObjective extends CraftScoreboardComponent implements Objective {
+public final class CraftObjective extends CraftScoreboardComponent implements Objective {
     private final ScoreboardObjective objective;
     private final CraftCriteria criteria;
 
@@ -19,7 +19,7 @@ final class CraftObjective extends CraftScoreboardComponent implements Objective
         this.criteria = CraftCriteria.getFromNMS(objective);
     }
 
-    ScoreboardObjective getHandle() {
+    public ScoreboardObjective getHandle() {
         return objective;
     }
 
diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScore.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScore.java
index d3ae91bc..3c57251c 100644
--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScore.java
+++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScore.java
@@ -17,7 +17,7 @@ import org.bukkit.scoreboard.Score;
  * <p>
  * Also, as an added perk, a CraftScore will (intentionally) stay a valid reference so long as objective is valid.
  */
-final class CraftScore implements Score {
+public final class CraftScore implements Score {
     private final String entry;
     private final CraftObjective objective;
 
diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboardComponent.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboardComponent.java
index d26d09d4..43b72ecf 100644
--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboardComponent.java
+++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboardComponent.java
@@ -1,6 +1,6 @@
 package org.bukkit.craftbukkit.scoreboard;
 
-abstract class CraftScoreboardComponent {
+public abstract class CraftScoreboardComponent {
     private CraftScoreboard scoreboard;
 
     CraftScoreboardComponent(CraftScoreboard scoreboard) {
diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboardTranslations.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboardTranslations.java
index d08e5a28..4a305bb1 100644
--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboardTranslations.java
+++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftScoreboardTranslations.java
@@ -6,7 +6,7 @@ import org.bukkit.scoreboard.DisplaySlot;
 
 import com.google.common.collect.ImmutableBiMap;
 
-class CraftScoreboardTranslations {
+public class CraftScoreboardTranslations {
     static final int MAX_DISPLAY_SLOT = 3;
     static ImmutableBiMap<DisplaySlot, String> SLOTS = ImmutableBiMap.of(
             DisplaySlot.BELOW_NAME, "belowName",
diff --git a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
index 22bf0655..2eed2a9d 100644
--- a/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
+++ b/src/main/java/org/bukkit/craftbukkit/scoreboard/CraftTeam.java
@@ -16,7 +16,7 @@ import net.minecraft.server.ScoreboardTeamBase;
 import org.bukkit.ChatColor;
 import org.bukkit.craftbukkit.util.CraftChatMessage;
 
-final class CraftTeam extends CraftScoreboardComponent implements Team {
+public final class CraftTeam extends CraftScoreboardComponent implements Team {
     private final ScoreboardTeam team;
 
     CraftTeam(CraftScoreboard scoreboard, ScoreboardTeam team) {
