From: Jedediah Smith <jedediah@silencegreys.com>
Date: Mon, 21 Mar 2016 01:42:17 -0400
Subject: [PATCH] Multi-passenger API


diff --git a/src/main/java/net/minecraft/server/Entity.java b/src/main/java/net/minecraft/server/Entity.java
--- a/src/main/java/net/minecraft/server/Entity.java
+++ b/src/main/java/net/minecraft/server/Entity.java
@@ -1904,6 +1904,7 @@ public abstract class Entity implements ICommandListener {
         return this.j <= 0;
     }
 
+    public void ejectAllPassengers() { this.ejectPassengers(); } // SportBukkit - alias for following method
     public void ejectPassengers() {
         for (int i = this.passengers.size() - 1; i >= 0; --i) {
             ((Entity) this.passengers.get(i)).stopRiding();
@@ -2750,6 +2751,7 @@ public abstract class Entity implements ICommandListener {
         return null;
     }
 
+    public List<Entity> getPassengers() { return bD(); } // SportBukkit - alias
     public List<Entity> bD() {
         return (List) (this.passengers.isEmpty() ? Collections.emptyList() : Lists.newArrayList(this.passengers));
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftEntity.java
@@ -2,12 +2,10 @@ package org.bukkit.craftbukkit.entity;
 
 import com.google.common.base.Function;
 import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Lists;
 
-import java.util.Collection;
-import java.util.List;
-import java.util.Set;
-import java.util.UUID;
+import java.util.*;
 
 import net.md_5.bungee.api.chat.BaseComponent;
 import net.minecraft.server.*;
@@ -395,16 +393,6 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
     }
 
     @Override
-    public List<org.bukkit.entity.Entity> getPassengers() {
-        return Lists.newArrayList(Lists.transform(getHandle().passengers, new Function<Entity, org.bukkit.entity.Entity>() {
-            @Override
-            public org.bukkit.entity.Entity apply(Entity input) {
-                return input.getBukkitEntity();
-            }
-        }));
-    }
-
-    @Override
     public boolean addPassenger(org.bukkit.entity.Entity passenger) {
         Preconditions.checkArgument(passenger != null, "passenger == null");
 
@@ -432,6 +420,47 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         return true;
     }
 
+    @Override
+    public boolean hasPassenger(org.bukkit.entity.Entity passenger) {
+        return equals(passenger.getVehicle());
+    }
+
+    @Override
+    public List<org.bukkit.entity.Entity> getPassengers() {
+        return Lists.transform(getHandle().getPassengers(), new Function<Entity, org.bukkit.entity.Entity>() {
+            @Override
+            public org.bukkit.entity.Entity apply(Entity passenger) {
+                return passenger.getBukkitEntity();
+            }
+        });
+    }
+
+    @Override
+    public List<org.bukkit.entity.Entity> setPassengers(List<org.bukkit.entity.Entity> passengers) {
+        for(org.bukkit.entity.Entity currentPassenger : ImmutableList.copyOf(getPassengers())) {
+            if(!passengers.contains(currentPassenger)) {
+                currentPassenger.leaveVehicle();
+            }
+        }
+
+        List<org.bukkit.entity.Entity> rejected = null;
+        for(org.bukkit.entity.Entity passenger : passengers) {
+            if(!passenger.enterVehicle(this)) {
+                if(rejected == null) {
+                    rejected = new ArrayList<org.bukkit.entity.Entity>();
+                }
+                rejected.add(passenger);
+            }
+        }
+
+        return rejected != null ? rejected : Collections.<org.bukkit.entity.Entity>emptyList();
+    }
+
+    @Override
+    public void ejectAll() {
+        getHandle().ejectAllPassengers();
+    }
+
     public float getFallDistance() {
         return getHandle().fallDistance;
     }
@@ -524,6 +553,12 @@ public abstract class CraftEntity implements org.bukkit.entity.Entity {
         return getHandle().isPassenger();
     }
 
+    @Override
+    public boolean enterVehicle(org.bukkit.entity.Entity vehicle) {
+        leaveVehicle();
+        return getHandle().startRiding(((CraftEntity) vehicle).getHandle());
+    }
+
     public boolean leaveVehicle() {
         if (!isInsideVehicle()) {
             return false;
