From 271ae5ad197195937b8f64fcc27a1b2219b4c7b3 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Thu, 6 Aug 2015 22:18:57 -0400
Subject: [PATCH] Skull skin API


diff --git a/src/main/java/org/bukkit/craftbukkit/block/CraftSkull.java b/src/main/java/org/bukkit/craftbukkit/block/CraftSkull.java
index f2865688..d4b7d54d 100644
--- a/src/main/java/org/bukkit/craftbukkit/block/CraftSkull.java
+++ b/src/main/java/org/bukkit/craftbukkit/block/CraftSkull.java
@@ -8,10 +8,15 @@ import org.bukkit.Bukkit;
 import org.bukkit.Material;
 import org.bukkit.OfflinePlayer;
 
+import org.bukkit.Skin;
 import org.bukkit.SkullType;
 import org.bukkit.block.Block;
 import org.bukkit.block.BlockFace;
 import org.bukkit.block.Skull;
+import org.bukkit.craftbukkit.CraftWorld;
+import org.bukkit.craftbukkit.util.Skins;
+
+import java.util.UUID;
 
 public class CraftSkull extends CraftBlockEntityState<TileEntitySkull> implements Skull {
 
@@ -207,6 +212,17 @@ public class CraftSkull extends CraftBlockEntityState<TileEntitySkull> implement
     }
 
     @Override
+    public void setOwner(String name, UUID uuid, Skin skin) {
+        profile = new GameProfile(uuid, name);
+        Skins.setProperties(skin, profile.getProperties());
+        skullType = SkullType.PLAYER;
+    }
+
+    @Override
+    public void clearOwner() {
+        profile = null;
+    }
+
     public BlockFace getRotation() {
         return getBlockFace(rotation);
     }
diff --git a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
index fbc16890..5f03e393 100644
--- a/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
+++ b/src/main/java/org/bukkit/craftbukkit/inventory/CraftMetaSkull.java
@@ -1,6 +1,7 @@
 package org.bukkit.craftbukkit.inventory;
 
 import java.util.Map;
+import java.util.UUID;
 
 import net.minecraft.server.GameProfileSerializer;
 import net.minecraft.server.NBTBase;
@@ -9,9 +10,11 @@ import net.minecraft.server.NBTTagCompound;
 import org.bukkit.Bukkit;
 import org.bukkit.Material;
 import org.bukkit.OfflinePlayer;
+import org.bukkit.Skin;
 import org.bukkit.configuration.serialization.DelegateDeserialization;
 import org.bukkit.craftbukkit.inventory.CraftMetaItem.SerializableMeta;
 import org.bukkit.craftbukkit.util.CraftMagicNumbers;
+import org.bukkit.craftbukkit.util.Skins;
 import org.bukkit.inventory.meta.SkullMeta;
 
 import com.google.common.collect.ImmutableMap.Builder;
@@ -149,6 +152,16 @@ class CraftMetaSkull extends CraftMetaItem implements SkullMeta {
         return true;
     }
 
+    public void setOwner(String name, UUID uuid, Skin skin) {
+        profile = new GameProfile(uuid, name);
+        Skins.setProperties(skin, profile.getProperties());
+    }
+
+    @Override
+    public void clearOwner() {
+        profile = null;
+    }
+
     @Override
     int applyHash() {
         final int original;
