From: Jedediah Smith <jedediah@silencegreys.com>
Date: Tue, 19 Apr 2016 20:30:55 -0400
Subject: [PATCH] Use client ticks to make attack speed more accurate


diff --git a/src/main/java/net/minecraft/server/EntityHuman.java b/src/main/java/net/minecraft/server/EntityHuman.java
--- a/src/main/java/net/minecraft/server/EntityHuman.java
+++ b/src/main/java/net/minecraft/server/EntityHuman.java
@@ -65,7 +65,9 @@ public abstract class EntityHuman extends EntityLiving {
     @Nullable
     public EntityFishingHook hookedFish;
 
-    Entity getRiding() { return this.bH(); } // SportBukkit - alias for following method@J+    // SportBukkit start - natural regen API
+    public int clientTicksSinceLastAttack; // SportBukkit
+
+    // SportBukkit start - natural regen API
     public boolean fastNaturalRegeneration = true;
     public boolean slowNaturalRegeneration = true;
     // SportBukkit end
@@ -1913,11 +1915,12 @@ public abstract class EntityHuman extends EntityLiving {
     }
 
     public float n(float f) {
-        return MathHelper.a(((float) this.aE + f) / this.dr(), 0.0F, 1.0F);
+        return MathHelper.a(((float) Math.max(this.aE, this.clientTicksSinceLastAttack) + f) / this.dr(), 0.0F, 1.0F); // SportBukkit - check client and server ticks
     }
 
     public void ds() {
         this.aE = 0;
+        this.clientTicksSinceLastAttack = 0; // SportBukkit
     }
 
     public ItemCooldown getCooldownTracker() {
diff --git a/src/main/java/net/minecraft/server/PlayerConnection.java b/src/main/java/net/minecraft/server/PlayerConnection.java
--- a/src/main/java/net/minecraft/server/PlayerConnection.java
+++ b/src/main/java/net/minecraft/server/PlayerConnection.java
@@ -528,6 +528,8 @@ public class PlayerConnection implements PacketListenerPlayIn, ITickable {
                     this.syncPosition();
                 }
 
+                this.player.clientTicksSinceLastAttack++; // SportBukkit
+
                 // SportBukkit start - velocity prediction
                 this.player.handleClientTick(
                         packetplayinflying.hasPos ? new Vector(packetplayinflying.x,
