--- a/net/minecraft/server/CommandTp.java
+++ b/net/minecraft/server/CommandTp.java
@@ -31,7 +31,7 @@
             if (astring.length != 2 && astring.length != 4 && astring.length != 6) {
                 object = a(icommandlistener);
             } else {
-                object = c(minecraftserver, icommandlistener, astring[0]);
+                object = b(minecraftserver, icommandlistener, astring[0]);
                 b0 = 1;
             }
 
@@ -51,19 +51,13 @@
                     a(icommandlistener, (ICommand) this, "commands.tp.success.coordinates", new Object[] { ((Entity) object).getName(), Double.valueOf(commandabstract_commandnumber.a()), Double.valueOf(commandabstract_commandnumber1.a()), Double.valueOf(commandabstract_commandnumber2.a())});
                 }
             } else {
-                Entity entity = c(minecraftserver, icommandlistener, astring[astring.length - 1]);
-
-                if (entity.world != ((Entity) object).world) {
-                    throw new CommandException("commands.tp.notSameDimension", new Object[0]);
-                } else {
-                    ((Entity) object).stopRiding();
-                    if (object instanceof EntityPlayer) {
-                        ((EntityPlayer) object).playerConnection.a(entity.locX, entity.locY, entity.locZ, entity.yaw, entity.pitch);
-                    } else {
-                        ((Entity) object).setPositionRotation(entity.locX, entity.locY, entity.locZ, entity.yaw, entity.pitch);
-                    }
+                Entity entity = b(minecraftserver, icommandlistener, astring[astring.length - 1]);
 
+                // CraftBukkit Start
+                // Use Bukkit teleport method in all cases. It has cross dimensional handling, events
+                if (((Entity) object).getBukkitEntity().teleport(entity.getBukkitEntity(), org.bukkit.event.player.PlayerTeleportEvent.TeleportCause.COMMAND)) {
                     a(icommandlistener, (ICommand) this, "commands.tp.success", new Object[] { ((Entity) object).getName(), entity.getName()});
+                    // CraftBukkit End
                 }
             }
         }
@@ -108,17 +102,17 @@
 
             entity.stopRiding();
             ((EntityPlayer) entity).playerConnection.a(commandabstract_commandnumber.b(), commandabstract_commandnumber1.b(), commandabstract_commandnumber2.b(), f, f1, enumset);
-            entity.setHeadRotation(f);
+            entity.h(f);
         } else {
             float f2 = (float) MathHelper.g(commandabstract_commandnumber3.a());
 
             f = (float) MathHelper.g(commandabstract_commandnumber4.a());
             f = MathHelper.a(f, -90.0F, 90.0F);
             entity.setPositionRotation(commandabstract_commandnumber.a(), commandabstract_commandnumber1.a(), commandabstract_commandnumber2.a(), f2, f);
-            entity.setHeadRotation(f2);
+            entity.h(f2);
         }
 
-        if (!(entity instanceof EntityLiving) || !((EntityLiving) entity).cN()) {
+        if (!(entity instanceof EntityLiving) || !((EntityLiving) entity).cH()) {
             entity.motY = 0.0D;
             entity.onGround = true;
         }
@@ -126,10 +120,17 @@
     }
 
     public List<String> tabComplete(MinecraftServer minecraftserver, ICommandListener icommandlistener, String[] astring, @Nullable BlockPosition blockposition) {
-        return astring.length != 1 && astring.length != 2 ? Collections.emptyList() : a(astring, minecraftserver.getPlayers());
+        return astring.length != 1 && astring.length != 2 ? Collections.<String>emptyList() : a(astring, minecraftserver.getPlayers()); // CraftBukkit - decompile error
     }
 
     public boolean isListStart(String[] astring, int i) {
         return i == 0;
     }
+
+    // CraftBukkit start - fix decompile error
+    @Override
+    public int compareTo(ICommand o) {
+        return a((ICommand) o);
+    }
+    // CraftBukkit end
 }
