--- a/net/minecraft/server/CommandGamerule.java
+++ b/net/minecraft/server/CommandGamerule.java
@@ -1,6 +1,5 @@
 package net.minecraft.server;
 
-import java.util.Collection;
 import java.util.Collections;
 import java.util.Iterator;
 import java.util.List;
@@ -23,7 +22,7 @@
     }
 
     public void execute(MinecraftServer minecraftserver, ICommandListener icommandlistener, String[] astring) throws CommandException {
-        GameRules gamerules = this.a(minecraftserver);
+        GameRules gamerules = icommandlistener.getWorld().getGameRules(); // CraftBukkit - Use current world
         String s = astring.length > 0 ? astring[0] : "";
         String s1 = astring.length > 1 ? a(astring, 1) : "";
 
@@ -79,10 +78,6 @@
                 if (gamerules.a(astring[0], GameRules.EnumGameRuleType.BOOLEAN_VALUE)) {
                     return a(astring, new String[] { "true", "false"});
                 }
-
-                if (gamerules.a(astring[0], GameRules.EnumGameRuleType.d)) {
-                    return a(astring, (Collection) minecraftserver.aL().d().keySet());
-                }
             }
 
             return Collections.emptyList();
@@ -92,4 +87,11 @@
     private GameRules a(MinecraftServer minecraftserver) {
         return minecraftserver.getWorldServer(0).getGameRules();
     }
+
+    // CraftBukkit start - fix decompile error
+    @Override
+    public int compareTo(ICommand o) {
+        return a((ICommand) o);
+    }
+    // CraftBukkit end
 }
