--- a/net/minecraft/server/ContainerHorse.java
+++ b/net/minecraft/server/ContainerHorse.java
@@ -1,11 +1,33 @@
 package net.minecraft.server;
 
+// CraftBukkit start
+import org.bukkit.craftbukkit.inventory.CraftInventory;
+import org.bukkit.craftbukkit.inventory.CraftInventoryView;
+import org.bukkit.inventory.InventoryView;
+// CraftBukkit end
+
 public class ContainerHorse extends Container {
 
     private final IInventory a;
     private final EntityHorseAbstract f;
 
+    // CraftBukkit start
+    org.bukkit.craftbukkit.inventory.CraftInventoryView bukkitEntity;
+    PlayerInventory player;
+
+    @Override
+    public InventoryView getBukkitView() {
+        if (bukkitEntity != null) {
+            return bukkitEntity;
+        }
+
+        CraftInventory inventory = new org.bukkit.craftbukkit.inventory.CraftInventoryHorse(this.a);
+        return bukkitEntity = new CraftInventoryView(player.player.getBukkitEntity(), inventory, this);
+    }
+
     public ContainerHorse(IInventory iinventory, final IInventory iinventory1, final EntityHorseAbstract entityhorseabstract, EntityHuman entityhuman) {
+        player = (PlayerInventory) iinventory;
+        // CraftBukkit end
         this.a = iinventory1;
         this.f = entityhorseabstract;
         boolean flag = true;
@@ -15,7 +37,7 @@
 
         this.a(new Slot(iinventory1, 0, 8, 18) {
             public boolean isAllowed(ItemStack itemstack) {
-                return itemstack.getItem() == Items.SADDLE && !this.hasItem() && entityhorseabstract.dD();
+                return itemstack.getItem() == Items.SADDLE && !this.hasItem() && entityhorseabstract.dA();
             }
         });
         this.a(new Slot(iinventory1, 1, 8, 36) {
@@ -32,8 +54,8 @@
 
         if (entityhorseabstract instanceof EntityHorseChestedAbstract && ((EntityHorseChestedAbstract) entityhorseabstract).isCarryingChest()) {
             for (i = 0; i < 3; ++i) {
-                for (j = 0; j < ((EntityHorseChestedAbstract) entityhorseabstract).dr(); ++j) {
-                    this.a(new Slot(iinventory1, 2 + j + i * ((EntityHorseChestedAbstract) entityhorseabstract).dr(), 80 + j * 18, 18 + i * 18));
+                for (j = 0; j < ((EntityHorseChestedAbstract) entityhorseabstract).dl(); ++j) {
+                    this.a(new Slot(iinventory1, 2 + j + i * ((EntityHorseChestedAbstract) entityhorseabstract).dl(), 80 + j * 18, 18 + i * 18));
                 }
             }
         }
