--- a/net/minecraft/server/CrashReport.java
+++ b/net/minecraft/server/CrashReport.java
@@ -2,13 +2,11 @@
 
 import com.google.common.collect.Lists;
 import java.io.File;
-import java.io.FileOutputStream;
-import java.io.OutputStreamWriter;
+import java.io.FileWriter;
 import java.io.PrintWriter;
 import java.io.StringWriter;
 import java.lang.management.ManagementFactory;
 import java.lang.management.RuntimeMXBean;
-import java.nio.charset.StandardCharsets;
 import java.text.SimpleDateFormat;
 import java.util.Date;
 import java.util.Iterator;
@@ -24,6 +22,7 @@
     private final String b;
     private final Throwable c;
     private final CrashReportSystemDetails d = new CrashReportSystemDetails(this, "System Details");
+    public List<CrashReportSystemDetails> extras() { return e; } // SportBukkit - alias
     private final List<CrashReportSystemDetails> e = Lists.newArrayList();
     private File f;
     private boolean g = true;
@@ -38,7 +37,7 @@
     private void h() {
         this.d.a("Minecraft Version", new CrashReportCallable() {
             public String a() {
-                return "1.12-pre6";
+                return "1.11.1";
             }
 
             public Object call() throws Exception {
@@ -125,12 +124,15 @@
                 return this.a();
             }
         });
+        this.d.a("CraftBukkit Information", (CrashReportCallable) new org.bukkit.craftbukkit.CraftCrashReport()); // CraftBukkit
     }
 
+    public String activity() { return a(); } // SportBukkit alias
     public String a() {
         return this.b;
     }
 
+    public Throwable exception() { return b(); } // SportBukkit - alias
     public Throwable b() {
         return this.c;
     }
@@ -234,28 +236,29 @@
                 file.getParentFile().mkdirs();
             }
 
-            OutputStreamWriter outputstreamwriter = null;
+            FileWriter filewriter = null;
 
             boolean flag;
 
             try {
-                outputstreamwriter = new OutputStreamWriter(new FileOutputStream(file), StandardCharsets.UTF_8);
-                outputstreamwriter.write(this.e());
+                filewriter = new FileWriter(file);
+                filewriter.write(this.e());
                 this.f = file;
                 boolean flag1 = true;
 
                 return flag1;
             } catch (Throwable throwable) {
-                CrashReport.a.error("Could not save crash report to {}", file, throwable);
+                CrashReport.a.error("Could not save crash report to {}", new Object[] { file, throwable});
                 flag = false;
             } finally {
-                IOUtils.closeQuietly(outputstreamwriter);
+                IOUtils.closeQuietly(filewriter);
             }
 
             return flag;
         }
     }
 
+    public CrashReportSystemDetails details() { return g(); } // SportBukkit - alias
     public CrashReportSystemDetails g() {
         return this.d;
     }
