--- a/net/minecraft/server/EntitySkeleton.java
+++ b/net/minecraft/server/EntitySkeleton.java
@@ -17,24 +17,24 @@
         return LootTables.ao;
     }
 
-    protected SoundEffect F() {
-        return SoundEffects.gQ;
+    protected SoundEffect G() {
+        return SoundEffects.fP;
     }
 
-    protected SoundEffect d(DamageSource damagesource) {
-        return SoundEffects.gV;
+    protected SoundEffect bW() {
+        return SoundEffects.fU;
     }
 
-    protected SoundEffect cd() {
-        return SoundEffects.gR;
+    protected SoundEffect bX() {
+        return SoundEffects.fQ;
     }
 
-    SoundEffect p() {
-        return SoundEffects.gX;
+    SoundEffect o() {
+        return SoundEffects.fW;
     }
 
     public void die(DamageSource damagesource) {
-        super.die(damagesource);
+        // super.die(damagesource); // CraftBukkit
         if (damagesource.getEntity() instanceof EntityCreeper) {
             EntityCreeper entitycreeper = (EntityCreeper) damagesource.getEntity();
 
@@ -43,6 +43,7 @@
                 this.a(new ItemStack(Items.SKULL, 1, 0), 0.0F);
             }
         }
+        super.die(damagesource); // CraftBukkit - moved from above
 
     }
 
