--- a/net/minecraft/server/EntityShulkerBullet.java
+++ b/net/minecraft/server/EntityShulkerBullet.java
@@ -30,7 +30,7 @@
         this.noclip = true;
     }
 
-    public SoundCategory bI() {
+    public SoundCategory bC() {
         return SoundCategory.HOSTILE;
     }
 
@@ -46,8 +46,29 @@
         this.target = entity;
         this.c = EnumDirection.UP;
         this.a(enumdirection_enumaxis);
+        projectileSource = (org.bukkit.entity.LivingEntity) entityliving.getBukkitEntity(); // CraftBukkit
     }
 
+    // CraftBukkit start
+    public EntityLiving getShooter() {
+        return this.shooter;
+    }
+
+    public void setShooter(EntityLiving e) {
+        this.shooter = e;
+    }
+
+    public Entity getTarget() {
+        return this.target;
+    }
+
+    public void setTarget(Entity e) {
+        this.target = e;
+        this.c = EnumDirection.UP;
+        this.a(EnumDirection.EnumAxis.X);
+    }
+    // CraftBukkit end
+
     protected void b(NBTTagCompound nbttagcompound) {
         BlockPosition blockposition;
         NBTTagCompound nbttagcompound1;
@@ -189,11 +210,11 @@
         this.d = 10 + this.random.nextInt(5) * 10;
     }
 
-    public void B_() {
+    public void A_() {
         if (!this.world.isClientSide && this.world.getDifficulty() == EnumDifficulty.PEACEFUL) {
             this.die();
         } else {
-            super.B_();
+            super.A_();
             if (!this.world.isClientSide) {
                 List list;
                 Iterator iterator;
@@ -282,14 +303,14 @@
         return false;
     }
 
-    public float f(float f) {
+    public float e(float f) {
         return 1.0F;
     }
 
     protected void a(MovingObjectPosition movingobjectposition) {
         if (movingobjectposition.entity == null) {
             ((WorldServer) this.world).a(EnumParticle.EXPLOSION_LARGE, this.locX, this.locY, this.locZ, 2, 0.2D, 0.2D, 0.2D, 0.0D, new int[0]);
-            this.a(SoundEffects.gD, 1.0F, 1.0F);
+            this.a(SoundEffects.fC, 1.0F, 1.0F);
         } else {
             boolean flag = movingobjectposition.entity.damageEntity(DamageSource.a(this, this.shooter).b(), 4.0F);
 
@@ -310,7 +331,7 @@
 
     public boolean damageEntity(DamageSource damagesource, float f) {
         if (!this.world.isClientSide) {
-            this.a(SoundEffects.gE, 1.0F, 1.0F);
+            this.a(SoundEffects.fD, 1.0F, 1.0F);
             ((WorldServer) this.world).a(EnumParticle.CRIT, this.locX, this.locY, this.locZ, 15, 0.2D, 0.2D, 0.2D, 0.0D, new int[0]);
             this.die();
         }
