--- a/net/minecraft/server/EntityEnderman.java
+++ b/net/minecraft/server/EntityEnderman.java
@@ -4,6 +4,8 @@
 import com.google.common.base.Optional;
 import com.google.common.base.Predicate;
 import com.google.common.collect.Sets;
+import org.bukkit.event.entity.EntityTargetEvent;
+
 import java.util.Random;
 import java.util.Set;
 import java.util.UUID;
@@ -14,10 +16,10 @@
     private static final UUID a = UUID.fromString("020E0DFB-87AE-4653-9556-831010E291A0");
     private static final AttributeModifier b = (new AttributeModifier(EntityEnderman.a, "Attacking speed boost", 0.15000000596046448D, 0)).a(false);
     private static final Set<Block> c = Sets.newIdentityHashSet();
-    private static final DataWatcherObject<Optional<IBlockData>> bx = DataWatcher.a(EntityEnderman.class, DataWatcherRegistry.g);
-    private static final DataWatcherObject<Boolean> by = DataWatcher.a(EntityEnderman.class, DataWatcherRegistry.h);
+    private static final DataWatcherObject<Optional<IBlockData>> bw = DataWatcher.a(EntityEnderman.class, DataWatcherRegistry.g);
+    private static final DataWatcherObject<Boolean> bx = DataWatcher.a(EntityEnderman.class, DataWatcherRegistry.h);
+    private int by;
     private int bz;
-    private int bA;
 
     public EntityEnderman(World world) {
         super(world);
@@ -38,7 +40,7 @@
         this.targetSelector.a(2, new PathfinderGoalHurtByTarget(this, false, new Class[0]));
         this.targetSelector.a(3, new PathfinderGoalNearestAttackableTarget(this, EntityEndermite.class, 10, true, false, new Predicate() {
             public boolean a(@Nullable EntityEndermite entityendermite) {
-                return entityendermite.p();
+                return entityendermite.o();
             }
 
             public boolean apply(@Nullable Object object) {
@@ -56,42 +58,53 @@
     }
 
     public void setGoalTarget(@Nullable EntityLiving entityliving) {
-        super.setGoalTarget(entityliving);
+        // CraftBukkit start - fire event
+        setGoalTarget(entityliving, EntityTargetEvent.TargetReason.UNKNOWN, true);
+    }
+
+    @Override
+    public boolean setGoalTarget(EntityLiving entityliving, org.bukkit.event.entity.EntityTargetEvent.TargetReason reason, boolean fireEvent) {
+        if (!super.setGoalTarget(entityliving, reason, fireEvent)) {
+            return false;
+        }
+        entityliving = getGoalTarget();
+        // CraftBukkit end
         AttributeInstance attributeinstance = this.getAttributeInstance(GenericAttributes.MOVEMENT_SPEED);
 
         if (entityliving == null) {
-            this.bA = 0;
-            this.datawatcher.set(EntityEnderman.by, Boolean.valueOf(false));
+            this.bz = 0;
+            this.datawatcher.set(EntityEnderman.bx, Boolean.valueOf(false));
             attributeinstance.c(EntityEnderman.b);
         } else {
-            this.bA = this.ticksLived;
-            this.datawatcher.set(EntityEnderman.by, Boolean.valueOf(true));
+            this.bz = this.ticksLived;
+            this.datawatcher.set(EntityEnderman.bx, Boolean.valueOf(true));
             if (!attributeinstance.a(EntityEnderman.b)) {
                 attributeinstance.b(EntityEnderman.b);
             }
         }
+        return true;
 
     }
 
     protected void i() {
         super.i();
-        this.datawatcher.register(EntityEnderman.bx, Optional.absent());
-        this.datawatcher.register(EntityEnderman.by, Boolean.valueOf(false));
+        this.datawatcher.register(EntityEnderman.bw, Optional.absent());
+        this.datawatcher.register(EntityEnderman.bx, Boolean.valueOf(false));
     }
 
-    public void p() {
-        if (this.ticksLived >= this.bz + 400) {
-            this.bz = this.ticksLived;
+    public void o() {
+        if (this.ticksLived >= this.by + 400) {
+            this.by = this.ticksLived;
             if (!this.isSilent()) {
-                this.world.a(this.locX, this.locY + (double) this.getHeadHeight(), this.locZ, SoundEffects.bh, this.bI(), 2.5F, 1.0F, false);
+                this.world.a(this.locX, this.locY + (double) this.getHeadHeight(), this.locZ, SoundEffects.bd, this.bC(), 2.5F, 1.0F, false);
             }
         }
 
     }
 
     public void a(DataWatcherObject<?> datawatcherobject) {
-        if (EntityEnderman.by.equals(datawatcherobject) && this.dm() && this.world.isClientSide) {
-            this.p();
+        if (EntityEnderman.bx.equals(datawatcherobject) && this.dk() && this.world.isClientSide) {
+            this.o();
         }
 
         super.a(datawatcherobject);
@@ -129,13 +142,13 @@
         this.setCarried(iblockdata);
     }
 
-    private boolean f(EntityHuman entityhuman) {
+    private boolean c(EntityHuman entityhuman) {
         ItemStack itemstack = (ItemStack) entityhuman.inventory.armor.get(3);
 
         if (itemstack.getItem() == Item.getItemOf(Blocks.PUMPKIN)) {
             return false;
         } else {
-            Vec3D vec3d = entityhuman.g(1.0F).a();
+            Vec3D vec3d = entityhuman.f(1.0F).a();
             Vec3D vec3d1 = new Vec3D(this.locX - entityhuman.locX, this.getBoundingBox().b + (double) this.getHeadHeight() - (entityhuman.locY + (double) entityhuman.getHeadHeight()), this.locZ - entityhuman.locZ);
             double d0 = vec3d1.b();
 
@@ -162,23 +175,23 @@
     }
 
     protected void M() {
-        if (this.an()) {
+        if (this.ai()) {
             this.damageEntity(DamageSource.DROWN, 1.0F);
         }
 
-        if (this.world.D() && this.ticksLived >= this.bA + 600) {
-            float f = this.f(1.0F);
+        if (this.world.B() && this.ticksLived >= this.bz + 600) {
+            float f = this.e(1.0F);
 
             if (f > 0.5F && this.world.h(new BlockPosition(this)) && this.random.nextFloat() * 30.0F < (f - 0.4F) * 2.0F) {
                 this.setGoalTarget((EntityLiving) null);
-                this.dk();
+                this.di();
             }
         }
 
         super.M();
     }
 
-    protected boolean dk() {
+    protected boolean di() {
         double d0 = this.locX + (this.random.nextDouble() - 0.5D) * 64.0D;
         double d1 = this.locY + (double) (this.random.nextInt(64) - 32);
         double d2 = this.locZ + (this.random.nextDouble() - 0.5D) * 64.0D;
@@ -202,23 +215,23 @@
         boolean flag = this.j(d0, d1, d2);
 
         if (flag) {
-            this.world.a((EntityHuman) null, this.lastX, this.lastY, this.lastZ, SoundEffects.bi, this.bI(), 1.0F, 1.0F);
-            this.a(SoundEffects.bi, 1.0F, 1.0F);
+            this.world.a((EntityHuman) null, this.lastX, this.lastY, this.lastZ, SoundEffects.be, this.bC(), 1.0F, 1.0F);
+            this.a(SoundEffects.be, 1.0F, 1.0F);
         }
 
         return flag;
     }
 
-    protected SoundEffect F() {
-        return this.dm() ? SoundEffects.bg : SoundEffects.bd;
+    protected SoundEffect G() {
+        return this.dk() ? SoundEffects.bc : SoundEffects.aZ;
     }
 
-    protected SoundEffect d(DamageSource damagesource) {
-        return SoundEffects.bf;
+    protected SoundEffect bW() {
+        return SoundEffects.bb;
     }
 
-    protected SoundEffect cd() {
-        return SoundEffects.be;
+    protected SoundEffect bX() {
+        return SoundEffects.ba;
     }
 
     protected void dropEquipment(boolean flag, int i) {
@@ -227,7 +240,7 @@
 
         if (iblockdata != null) {
             Item item = Item.getItemOf(iblockdata.getBlock());
-            int j = item.k() ? iblockdata.getBlock().toLegacyData(iblockdata) : 0;
+            int j = item.l() ? iblockdata.getBlock().toLegacyData(iblockdata) : 0;
 
             this.a(new ItemStack(item, 1, j), 0.0F);
         }
@@ -240,12 +253,12 @@
     }
 
     public void setCarried(@Nullable IBlockData iblockdata) {
-        this.datawatcher.set(EntityEnderman.bx, Optional.fromNullable(iblockdata));
+        this.datawatcher.set(EntityEnderman.bw, Optional.fromNullable(iblockdata));
     }
 
     @Nullable
     public IBlockData getCarried() {
-        return (IBlockData) ((Optional) this.datawatcher.get(EntityEnderman.bx)).orNull();
+        return (IBlockData) ((Optional) this.datawatcher.get(EntityEnderman.bw)).orNull();
     }
 
     public boolean damageEntity(DamageSource damagesource, float f) {
@@ -253,7 +266,7 @@
             return false;
         } else if (damagesource instanceof EntityDamageSourceIndirect) {
             for (int i = 0; i < 64; ++i) {
-                if (this.dk()) {
+                if (this.di()) {
                     return true;
                 }
             }
@@ -263,15 +276,15 @@
             boolean flag = super.damageEntity(damagesource, f);
 
             if (damagesource.ignoresArmor() && this.random.nextInt(10) != 0) {
-                this.dk();
+                this.di();
             }
 
             return flag;
         }
     }
 
-    public boolean dm() {
-        return ((Boolean) this.datawatcher.get(EntityEnderman.by)).booleanValue();
+    public boolean dk() {
+        return ((Boolean) this.datawatcher.get(EntityEnderman.bx)).booleanValue();
     }
 
     static {
@@ -317,8 +330,12 @@
             boolean flag = movingobjectposition != null && movingobjectposition.a().equals(blockposition);
 
             if (EntityEnderman.c.contains(block) && flag) {
-                this.enderman.setCarried(iblockdata);
-                world.setAir(blockposition);
+                // CraftBukkit start - Pickup event
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.enderman, this.enderman.world.getWorld().getBlockAt(blockposition.getX(), blockposition.getY(), blockposition.getZ()), org.bukkit.Material.AIR).isCancelled()) {
+                    this.enderman.setCarried(iblockdata);
+                    world.setAir(blockposition);
+                }
+                // CraftBukkit end
             }
 
         }
@@ -348,14 +365,18 @@
             IBlockData iblockdata2 = this.a.getCarried();
 
             if (iblockdata2 != null && this.a(world, blockposition, iblockdata2.getBlock(), iblockdata, iblockdata1)) {
+                // CraftBukkit start - Place event
+                if (!org.bukkit.craftbukkit.event.CraftEventFactory.callEntityChangeBlockEvent(this.a, blockposition, this.a.getCarried().getBlock(), this.a.getCarried().getBlock().toLegacyData(this.a.getCarried())).isCancelled()) {
                 world.setTypeAndData(blockposition, iblockdata2, 3);
                 this.a.setCarried((IBlockData) null);
+                }
+                // CraftBukkit end
             }
 
         }
 
         private boolean a(World world, BlockPosition blockposition, Block block, IBlockData iblockdata, IBlockData iblockdata1) {
-            return !block.canPlace(world, blockposition) ? false : (iblockdata.getMaterial() != Material.AIR ? false : (iblockdata1.getMaterial() == Material.AIR ? false : iblockdata1.g()));
+            return !block.canPlace(world, blockposition) ? false : (iblockdata.getMaterial() != Material.AIR ? false : (iblockdata1.getMaterial() == Material.AIR ? false : iblockdata1.h()));
         }
     }
 
@@ -376,7 +397,7 @@
 
             this.j = this.i.world.a(this.i.locX, this.i.locY, this.i.locZ, d0, d0, (Function) null, new Predicate() {
                 public boolean a(@Nullable EntityHuman entityhuman) {
-                    return entityhuman != null && PathfinderGoalPlayerWhoLookedAtTarget.this.i.f(entityhuman);
+                    return entityhuman != null && PathfinderGoalPlayerWhoLookedAtTarget.this.i.c(entityhuman);
                 }
 
                 public boolean apply(@Nullable Object object) {
@@ -398,7 +419,7 @@
 
         public boolean b() {
             if (this.j != null) {
-                if (!this.i.f(this.j)) {
+                if (!this.i.c(this.j)) {
                     return false;
                 } else {
                     this.i.a((Entity) this.j, 10.0F, 10.0F);
@@ -418,9 +439,9 @@
                 }
             } else {
                 if (this.d != null) {
-                    if (this.i.f((EntityHuman) this.d)) {
+                    if (this.i.c((EntityHuman) this.d)) {
                         if (((EntityHuman) this.d).h(this.i) < 16.0D) {
-                            this.i.dk();
+                            this.i.di();
                         }
 
                         this.l = 0;
