--- a/net/minecraft/server/EntityLlama.java
+++ b/net/minecraft/server/EntityLlama.java
@@ -4,40 +4,59 @@
 import java.util.Iterator;
 import javax.annotation.Nullable;
 
+// SportBukkit start
+import com.google.common.collect.Range;
+import org.bukkit.craftbukkit.protocol.MetadataKey;
+// SportBukkit end
+
 public class EntityLlama extends EntityHorseChestedAbstract implements IRangedEntity {
 
-    private static final DataWatcherObject<Integer> bH = DataWatcher.a(EntityLlama.class, DataWatcherRegistry.b);
-    private static final DataWatcherObject<Integer> bI = DataWatcher.a(EntityLlama.class, DataWatcherRegistry.b);
-    private static final DataWatcherObject<Integer> bJ = DataWatcher.a(EntityLlama.class, DataWatcherRegistry.b);
-    private boolean bK;
+    // SportBukkit start
+    private static final DataWatcherObject<Integer> bG = MetadataKey
+        .builder(EntityLlama.class, DataWatcherRegistry.b)
+        .id(Range.atLeast(301), 16)
+        .build();
+
+    private static final DataWatcherObject<Integer> bH = MetadataKey
+        .builder(EntityLlama.class, DataWatcherRegistry.b)
+        .id(Range.atLeast(301), 17)
+        .build();
+
+    private static final DataWatcherObject<Integer> bI = MetadataKey
+        .builder(EntityLlama.class, DataWatcherRegistry.b)
+        .id(Range.atLeast(301), 18)
+        .build();
+    // SportBukkit end
+
+    private boolean bJ;
     @Nullable
-    private EntityLlama bL;
+    private EntityLlama bK;
     @Nullable
-    private EntityLlama bM;
+    private EntityLlama bL;
 
     public EntityLlama(World world) {
         super(world);
         this.setSize(0.9F, 1.87F);
     }
 
-    public void setStrength(int i) {
-        this.datawatcher.set(EntityLlama.bH, Integer.valueOf(Math.max(1, Math.min(5, i))));
+    public void p(int i) { // PAIL: private-> public, rename
+        this.datawatcher.set(EntityLlama.bG, Integer.valueOf(Math.max(1, Math.min(5, i))));
     }
 
-    private void dW() {
+    private void dT() {
         int i = this.random.nextFloat() < 0.04F ? 5 : 3;
 
-        this.setStrength(1 + this.random.nextInt(i));
+        this.p(1 + this.random.nextInt(i));
     }
 
-    public int getStrength() {
-        return ((Integer) this.datawatcher.get(EntityLlama.bH)).intValue();
+    public int dL() { // PAIL: rename
+        return ((Integer) this.datawatcher.get(EntityLlama.bG)).intValue();
     }
 
     public void b(NBTTagCompound nbttagcompound) {
         super.b(nbttagcompound);
         nbttagcompound.setInt("Variant", this.getVariant());
-        nbttagcompound.setInt("Strength", this.getStrength());
+        nbttagcompound.setInt("Strength", this.dL());
         if (!this.inventoryChest.getItem(1).isEmpty()) {
             nbttagcompound.set("DecorItem", this.inventoryChest.getItem(1).save(new NBTTagCompound()));
         }
@@ -45,14 +64,14 @@
     }
 
     public void a(NBTTagCompound nbttagcompound) {
-        this.setStrength(nbttagcompound.getInt("Strength"));
+        this.p(nbttagcompound.getInt("Strength"));
         super.a(nbttagcompound);
         this.setVariant(nbttagcompound.getInt("Variant"));
         if (nbttagcompound.hasKeyOfType("DecorItem", 10)) {
             this.inventoryChest.setItem(1, new ItemStack(nbttagcompound.getCompound("DecorItem")));
         }
 
-        this.dB();
+        this.dy();
     }
 
     protected void r() {
@@ -77,21 +96,21 @@
 
     protected void i() {
         super.i();
-        this.datawatcher.register(EntityLlama.bH, Integer.valueOf(0));
-        this.datawatcher.register(EntityLlama.bI, Integer.valueOf(-1));
-        this.datawatcher.register(EntityLlama.bJ, Integer.valueOf(0));
+        this.datawatcher.register(EntityLlama.bG, Integer.valueOf(0));
+        this.datawatcher.register(EntityLlama.bH, Integer.valueOf(-1));
+        this.datawatcher.register(EntityLlama.bI, Integer.valueOf(0));
     }
 
     public int getVariant() {
-        return MathHelper.clamp(((Integer) this.datawatcher.get(EntityLlama.bJ)).intValue(), 0, 3);
+        return MathHelper.clamp(((Integer) this.datawatcher.get(EntityLlama.bI)).intValue(), 0, 3);
     }
 
     public void setVariant(int i) {
-        this.datawatcher.set(EntityLlama.bJ, Integer.valueOf(i));
+        this.datawatcher.set(EntityLlama.bI, Integer.valueOf(i));
     }
 
-    protected int dl() {
-        return this.isCarryingChest() ? 2 + 3 * this.dr() : super.dl();
+    protected int di() {
+        return this.isCarryingChest() ? 2 + 3 * this.dl() : super.di();
     }
 
     public void k(Entity entity) {
@@ -100,15 +119,15 @@
             float f1 = MathHelper.sin(this.aN * 0.017453292F);
             float f2 = 0.3F;
 
-            entity.setPosition(this.locX + (double) (0.3F * f1), this.locY + this.aE() + entity.aD(), this.locZ - (double) (0.3F * f));
+            entity.setPosition(this.locX + (double) (0.3F * f1), this.locY + this.ay() + entity.ax(), this.locZ - (double) (0.3F * f));
         }
     }
 
-    public double aE() {
+    public double ay() {
         return (double) this.length * 0.67D;
     }
 
-    public boolean cT() {
+    public boolean cR() {
         return false;
     }
 
@@ -129,7 +148,7 @@
             f = 10.0F;
             if (this.isTamed() && this.getAge() == 0) {
                 flag = true;
-                this.f(entityhuman);
+                this.c(entityhuman);
             }
         }
 
@@ -155,21 +174,21 @@
         }
 
         if (flag && !this.isSilent()) {
-            this.world.a((EntityHuman) null, this.locX, this.locY, this.locZ, SoundEffects.dQ, this.bI(), 1.0F, 1.0F + (this.random.nextFloat() - this.random.nextFloat()) * 0.2F);
+            this.world.a((EntityHuman) null, this.locX, this.locY, this.locZ, SoundEffects.dD, this.bC(), 1.0F, 1.0F + (this.random.nextFloat() - this.random.nextFloat()) * 0.2F);
         }
 
         return flag;
     }
 
     protected boolean isFrozen() {
-        return this.getHealth() <= 0.0F || this.dw();
+        return this.getHealth() <= 0.0F || this.dt();
     }
 
     @Nullable
     public GroupDataEntity prepare(DifficultyDamageScaler difficultydamagescaler, @Nullable GroupDataEntity groupdataentity) {
         Object object = super.prepare(difficultydamagescaler, groupdataentity);
 
-        this.dW();
+        this.dT();
         int i;
 
         if (object instanceof EntityLlama.b) {
@@ -183,35 +202,35 @@
         return (GroupDataEntity) object;
     }
 
-    protected SoundEffect dm() {
-        return SoundEffects.dN;
+    protected SoundEffect dj() {
+        return SoundEffects.dA;
     }
 
-    protected SoundEffect F() {
-        return SoundEffects.dM;
+    protected SoundEffect G() {
+        return SoundEffects.dz;
     }
 
-    protected SoundEffect d(DamageSource damagesource) {
-        return SoundEffects.dR;
+    protected SoundEffect bW() {
+        return SoundEffects.dE;
     }
 
-    protected SoundEffect cd() {
-        return SoundEffects.dP;
+    protected SoundEffect bX() {
+        return SoundEffects.dC;
     }
 
     protected void a(BlockPosition blockposition, Block block) {
-        this.a(SoundEffects.dT, 0.15F, 1.0F);
+        this.a(SoundEffects.dG, 0.15F, 1.0F);
     }
 
-    protected void dn() {
-        this.a(SoundEffects.dO, 1.0F, (this.random.nextFloat() - this.random.nextFloat()) * 0.2F + 1.0F);
+    protected void dk() {
+        this.a(SoundEffects.dB, 1.0F, (this.random.nextFloat() - this.random.nextFloat()) * 0.2F + 1.0F);
     }
 
-    public void dI() {
-        SoundEffect soundeffect = this.dm();
+    public void dF() {
+        SoundEffect soundeffect = this.dj();
 
         if (soundeffect != null) {
-            this.a(soundeffect, this.co(), this.cp());
+            this.a(soundeffect, this.ci(), this.cj());
         }
 
     }
@@ -221,11 +240,11 @@
         return LootTables.aw;
     }
 
-    public int dr() {
-        return this.getStrength();
+    public int dl() {
+        return this.dL();
     }
 
-    public boolean dN() {
+    public boolean dK() {
         return true;
     }
 
@@ -233,31 +252,31 @@
         return itemstack.getItem() == Item.getItemOf(Blocks.CARPET);
     }
 
-    public boolean dD() {
+    public boolean dA() {
         return false;
     }
 
     public void a(IInventory iinventory) {
-        EnumColor enumcolor = this.dR();
+        EnumColor enumcolor = this.dO();
 
         super.a(iinventory);
-        EnumColor enumcolor1 = this.dR();
+        EnumColor enumcolor1 = this.dO();
 
         if (this.ticksLived > 20 && enumcolor1 != null && enumcolor1 != enumcolor) {
-            this.a(SoundEffects.dU, 0.5F, 1.0F);
+            this.a(SoundEffects.dH, 0.5F, 1.0F);
         }
 
     }
 
-    protected void dB() {
+    protected void dy() {
         if (!this.world.isClientSide) {
-            super.dB();
+            super.dy();
             this.g(this.inventoryChest.getItem(1));
         }
     }
 
     private void a(@Nullable EnumColor enumcolor) {
-        this.datawatcher.set(EntityLlama.bI, Integer.valueOf(enumcolor == null ? -1 : enumcolor.getColorIndex()));
+        this.datawatcher.set(EntityLlama.bH, Integer.valueOf(enumcolor == null ? -1 : enumcolor.getColorIndex()));
     }
 
     private void g(ItemStack itemstack) {
@@ -270,8 +289,8 @@
     }
 
     @Nullable
-    public EnumColor dR() {
-        int i = ((Integer) this.datawatcher.get(EntityLlama.bI)).intValue();
+    public EnumColor dO() {
+        int i = ((Integer) this.datawatcher.get(EntityLlama.bH)).intValue();
 
         return i == -1 ? null : EnumColor.fromColorIndex(i);
     }
@@ -281,7 +300,7 @@
     }
 
     public boolean mate(EntityAnimal entityanimal) {
-        return entityanimal != this && entityanimal instanceof EntityLlama && this.dJ() && ((EntityLlama) entityanimal).dJ();
+        return entityanimal != this && entityanimal instanceof EntityLlama && this.dG() && ((EntityLlama) entityanimal).dG();
     }
 
     public EntityLlama b(EntityAgeable entityageable) {
@@ -289,13 +308,13 @@
 
         this.a(entityageable, (EntityHorseAbstract) entityllama);
         EntityLlama entityllama1 = (EntityLlama) entityageable;
-        int i = this.random.nextInt(Math.max(this.getStrength(), entityllama1.getStrength())) + 1;
+        int i = this.random.nextInt(Math.max(this.dL(), entityllama1.dL())) + 1;
 
         if (this.random.nextFloat() < 0.03F) {
             ++i;
         }
 
-        entityllama.setStrength(i);
+        entityllama.p(i);
         entityllama.setVariant(this.random.nextBoolean() ? this.getVariant() : entityllama1.getVariant());
         return entityllama;
     }
@@ -308,13 +327,13 @@
         float f = MathHelper.sqrt(d0 * d0 + d2 * d2) * 0.2F;
 
         entityllamaspit.shoot(d0, d1 + (double) f, d2, 1.5F, 10.0F);
-        this.world.a((EntityHuman) null, this.locX, this.locY, this.locZ, SoundEffects.dS, this.bI(), 1.0F, 1.0F + (this.random.nextFloat() - this.random.nextFloat()) * 0.2F);
+        this.world.a((EntityHuman) null, this.locX, this.locY, this.locZ, SoundEffects.dF, this.bC(), 1.0F, 1.0F + (this.random.nextFloat() - this.random.nextFloat()) * 0.2F);
         this.world.addEntity(entityllamaspit);
-        this.bK = true;
+        this.bJ = true;
     }
 
-    private void y(boolean flag) {
-        this.bK = flag;
+    private void x(boolean flag) {
+        this.bJ = flag;
     }
 
     public void e(float f, float f1) {
@@ -324,7 +343,7 @@
             if (f >= 6.0F) {
                 this.damageEntity(DamageSource.FALL, (float) i);
                 if (this.isVehicle()) {
-                    Iterator iterator = this.bE().iterator();
+                    Iterator iterator = this.by().iterator();
 
                     while (iterator.hasNext()) {
                         Entity entity = (Entity) iterator.next();
@@ -340,50 +359,50 @@
             if (iblockdata.getMaterial() != Material.AIR && !this.isSilent()) {
                 SoundEffectType soundeffecttype = block.getStepSound();
 
-                this.world.a((EntityHuman) null, this.locX, this.locY, this.locZ, soundeffecttype.d(), this.bI(), soundeffecttype.a() * 0.5F, soundeffecttype.b() * 0.75F);
+                this.world.a((EntityHuman) null, this.locX, this.locY, this.locZ, soundeffecttype.d(), this.bC(), soundeffecttype.a() * 0.5F, soundeffecttype.b() * 0.75F);
             }
 
         }
     }
 
-    public void dS() {
-        if (this.bL != null) {
-            this.bL.bM = null;
+    public void dP() {
+        if (this.bK != null) {
+            this.bK.bL = null;
         }
 
-        this.bL = null;
+        this.bK = null;
     }
 
     public void a(EntityLlama entityllama) {
-        this.bL = entityllama;
-        this.bL.bM = this;
+        this.bK = entityllama;
+        this.bK.bL = this;
     }
 
-    public boolean dT() {
-        return this.bM != null;
+    public boolean dQ() {
+        return this.bL != null;
     }
 
-    public boolean dU() {
-        return this.bL != null;
+    public boolean dR() {
+        return this.bK != null;
     }
 
     @Nullable
-    public EntityLlama dV() {
-        return this.bL;
+    public EntityLlama dS() {
+        return this.bK;
     }
 
-    protected double di() {
+    protected double dg() {
         return 2.0D;
     }
 
-    protected void dG() {
-        if (!this.dU() && this.isBaby()) {
-            super.dG();
+    protected void dD() {
+        if (!this.dR() && this.isBaby()) {
+            super.dD();
         }
 
     }
 
-    public boolean dH() {
+    public boolean dE() {
         return false;
     }
 
@@ -391,8 +410,6 @@
         this.e(entityliving);
     }
 
-    public void p(boolean flag) {}
-
     public EntityAgeable createChild(EntityAgeable entityageable) {
         return this.b(entityageable);
     }
@@ -427,8 +444,8 @@
             if (this.e instanceof EntityLlama) {
                 EntityLlama entityllama = (EntityLlama) this.e;
 
-                if (entityllama.bK) {
-                    entityllama.y(false);
+                if (entityllama.bJ) {
+                    entityllama.x(false);
                     return false;
                 }
             }
