--- a/net/minecraft/server/EntityLlamaSpit.java
+++ b/net/minecraft/server/EntityLlamaSpit.java
@@ -7,7 +7,7 @@
 
 public class EntityLlamaSpit extends Entity implements IProjectile {
 
-    public EntityLlama shooter;
+    public EntityLiving shooter; // CraftBukkit - type
     private NBTTagCompound b;
 
     public EntityLlamaSpit(World world) {
@@ -21,8 +21,8 @@
         this.setSize(0.25F, 0.25F);
     }
 
-    public void B_() {
-        super.B_();
+    public void A_() {
+        super.A_();
         if (this.b != null) {
             this.j();
         }
