--- a/net/minecraft/server/EntityMushroomCow.java
+++ b/net/minecraft/server/EntityMushroomCow.java
@@ -1,13 +1,14 @@
 package net.minecraft.server;
 
 import javax.annotation.Nullable;
+import org.bukkit.event.player.PlayerShearEntityEvent; // CraftBukkit
 
 public class EntityMushroomCow extends EntityCow {
 
     public EntityMushroomCow(World world) {
         super(world);
         this.setSize(0.9F, 1.4F);
-        this.bA = Blocks.MYCELIUM;
+        this.bz = Blocks.MYCELIUM;
     }
 
     public static void c(DataConverterManager dataconvertermanager) {
@@ -27,6 +28,14 @@
 
             return true;
         } else if (itemstack.getItem() == Items.SHEARS && this.getAge() >= 0) {
+            // CraftBukkit start
+            PlayerShearEntityEvent event = new PlayerShearEntityEvent((org.bukkit.entity.Player) entityhuman.getBukkitEntity(), this.getBukkitEntity());
+            this.world.getServer().getPluginManager().callEvent(event);
+
+            if (event.isCancelled()) {
+                return false;
+            }
+            // CraftBukkit end
             this.die();
             this.world.addParticle(EnumParticle.EXPLOSION_LARGE, this.locX, this.locY + (double) (this.length / 2.0F), this.locZ, 0.0D, 0.0D, 0.0D, new int[0]);
             if (!this.world.isClientSide) {
@@ -46,7 +55,7 @@
                 }
 
                 itemstack.damage(1, entityhuman);
-                this.a(SoundEffects.ei, 1.0F, 1.0F);
+                this.a(SoundEffects.dV, 1.0F, 1.0F);
             }
 
             return true;
