--- a/net/minecraft/server/EntityEvoker.java
+++ b/net/minecraft/server/EntityEvoker.java
@@ -4,9 +4,12 @@
 import java.util.List;
 import javax.annotation.Nullable;
 
-public class EntityEvoker extends EntityIllagerWizard {
+public class EntityEvoker extends EntityMonster {
 
-    private EntitySheep c;
+    protected static final DataWatcherObject<Byte> a = DataWatcher.a(EntityEvoker.class, DataWatcherRegistry.a);
+    private int b;
+    private int c;
+    private EntitySheep bw;
 
     public EntityEvoker(World world) {
         super(world);
@@ -17,11 +20,11 @@
     protected void r() {
         super.r();
         this.goalSelector.a(0, new PathfinderGoalFloat(this));
-        this.goalSelector.a(1, new EntityEvoker.b(null));
+        this.goalSelector.a(1, new EntityEvoker.b());
         this.goalSelector.a(2, new PathfinderGoalAvoidTarget(this, EntityHuman.class, 8.0F, 0.6D, 1.0D));
         this.goalSelector.a(4, new EntityEvoker.c(null));
         this.goalSelector.a(5, new EntityEvoker.a(null));
-        this.goalSelector.a(6, new EntityEvoker.d());
+        this.goalSelector.a(6, new EntityEvoker.e());
         this.goalSelector.a(8, new PathfinderGoalRandomStroll(this, 0.6D));
         this.goalSelector.a(9, new PathfinderGoalLookAtPlayer(this, EntityHuman.class, 3.0F, 1.0F));
         this.goalSelector.a(10, new PathfinderGoalLookAtPlayer(this, EntityInsentient.class, 8.0F));
@@ -40,6 +43,7 @@
 
     protected void i() {
         super.i();
+        this.datawatcher.register(EntityEvoker.a, Byte.valueOf((byte) 0));
     }
 
     public static void a(DataConverterManager dataconvertermanager) {
@@ -48,54 +52,102 @@
 
     public void a(NBTTagCompound nbttagcompound) {
         super.a(nbttagcompound);
+        this.b = nbttagcompound.getInt("SpellTicks");
     }
 
     public void b(NBTTagCompound nbttagcompound) {
         super.b(nbttagcompound);
+        nbttagcompound.setInt("SpellTicks", this.b);
+    }
+
+    public EnumMonsterType getMonsterType() {
+        return EnumMonsterType.ILLAGER;
     }
 
     protected MinecraftKey J() {
         return LootTables.au;
     }
 
+    public boolean o() {
+        return this.world.isClientSide ? ((Byte) this.datawatcher.get(EntityEvoker.a)).byteValue() > 0 : this.b > 0;
+    }
+
+    public void a(int i) {
+        this.datawatcher.set(EntityEvoker.a, Byte.valueOf((byte) i));
+    }
+
+    // CraftBukkit start
+    public int getSpell() {
+        return this.datawatcher.get(EntityEvoker.a).intValue();
+    }
+    // CraftBukkit end
+
+    private int di() {
+        return this.b;
+    }
+
     protected void M() {
         super.M();
+        if (this.b > 0) {
+            --this.b;
+        }
+
     }
 
-    public void B_() {
-        super.B_();
+    public void A_() {
+        super.A_();
+        if (this.world.isClientSide && this.o()) {
+            byte b0 = ((Byte) this.datawatcher.get(EntityEvoker.a)).byteValue();
+            double d0 = 0.7D;
+            double d1 = 0.5D;
+            double d2 = 0.2D;
+
+            if (b0 == 2) {
+                d0 = 0.4D;
+                d1 = 0.3D;
+                d2 = 0.35D;
+            } else if (b0 == 1) {
+                d0 = 0.7D;
+                d1 = 0.7D;
+                d2 = 0.8D;
+            }
+
+            float f = this.aN * 0.017453292F + MathHelper.cos((float) this.ticksLived * 0.6662F) * 0.25F;
+            float f1 = MathHelper.cos(f);
+            float f2 = MathHelper.sin(f);
+
+            this.world.addParticle(EnumParticle.SPELL_MOB, this.locX + (double) f1 * 0.6D, this.locY + 1.8D, this.locZ + (double) f2 * 0.6D, d0, d1, d2, new int[0]);
+            this.world.addParticle(EnumParticle.SPELL_MOB, this.locX - (double) f1 * 0.6D, this.locY + 1.8D, this.locZ - (double) f2 * 0.6D, d0, d1, d2, new int[0]);
+        }
+
     }
 
     public boolean r(Entity entity) {
-        return entity == null ? false : (entity == this ? true : (super.r(entity) ? true : (entity instanceof EntityVex ? this.r(((EntityVex) entity).p()) : (entity instanceof EntityLiving && ((EntityLiving) entity).getMonsterType() == EnumMonsterType.ILLAGER ? this.aW() == null && entity.aW() == null : false))));
+        return entity == null ? false : (entity == this ? true : (super.r(entity) ? true : (entity instanceof EntityVex ? this.r(((EntityVex) entity).o()) : (entity instanceof EntityLiving && ((EntityLiving) entity).getMonsterType() == EnumMonsterType.ILLAGER ? this.aQ() == null && entity.aQ() == null : false))));
     }
 
-    protected SoundEffect F() {
-        return SoundEffects.bs;
+    protected SoundEffect G() {
+        return SoundEffects.bm;
     }
 
-    protected SoundEffect cd() {
-        return SoundEffects.bu;
+    protected SoundEffect bX() {
+        return SoundEffects.bo;
     }
 
-    protected SoundEffect d(DamageSource damagesource) {
-        return SoundEffects.bv;
+    protected SoundEffect bW() {
+        return SoundEffects.bp;
     }
 
     private void a(@Nullable EntitySheep entitysheep) {
-        this.c = entitysheep;
+        this.bw = entitysheep;
     }
 
     @Nullable
-    private EntitySheep do() {
-        return this.c;
+    private EntitySheep dj() {
+        return this.bw;
     }
 
-    protected SoundEffect dk() {
-        return SoundEffects.bt;
-    }
-
-    public class d extends EntityIllagerWizard.c {
+    public class e extends EntityEvoker.d {
 
         final Predicate<EntitySheep> a = new Predicate() {
             public boolean a(EntitySheep entitysheep) {
@@ -107,16 +159,16 @@
             }
         };
 
-        public d() {
-            super();
+        public e() {
+            super(null);
         }
 
         public boolean a() {
             if (EntityEvoker.this.getGoalTarget() != null) {
                 return false;
-            } else if (EntityEvoker.this.dl()) {
+            } else if (EntityEvoker.this.o()) {
                 return false;
-            } else if (EntityEvoker.this.ticksLived < this.d) {
+            } else if (EntityEvoker.this.ticksLived < this.c) {
                 return false;
             } else if (!EntityEvoker.this.world.getGameRules().getBoolean("mobGriefing")) {
                 return false;
@@ -133,7 +185,7 @@
         }
 
         public boolean b() {
-            return EntityEvoker.this.do() != null && this.c > 0;
+            return EntityEvoker.this.dj() != null && this.b > 0;
         }
 
         public void d() {
@@ -142,7 +194,7 @@
         }
 
         protected void j() {
-            EntitySheep entitysheep = EntityEvoker.this.do();
+            EntitySheep entitysheep = EntityEvoker.this.dj();
 
             if (entitysheep != null && entitysheep.isAlive()) {
                 entitysheep.setColor(EnumColor.RED);
@@ -163,18 +215,18 @@
         }
 
         protected SoundEffect k() {
-            return SoundEffects.by;
+            return SoundEffects.bs;
         }
 
-        protected EntityIllagerWizard.Spell l() {
-            return EntityIllagerWizard.Spell.WOLOLO;
+        protected int l() {
+            return 3;
         }
     }
 
-    class c extends EntityIllagerWizard.c {
+    class c extends EntityEvoker.d {
 
         private c() {
-            super();
+            super(null);
         }
 
         public boolean a() {
@@ -211,11 +263,11 @@
         }
 
         protected SoundEffect k() {
-            return SoundEffects.bx;
+            return SoundEffects.br;
         }
 
-        protected EntityIllagerWizard.Spell l() {
-            return EntityIllagerWizard.Spell.SUMMON_VEX;
+        protected int l() {
+            return 1;
         }
 
         c(Object object) {
@@ -223,10 +275,10 @@
         }
     }
 
-    class a extends EntityIllagerWizard.c {
+    class a extends EntityEvoker.d {
 
         private a() {
-            super();
+            super(null);
         }
 
         protected int f() {
@@ -244,7 +296,7 @@
             float f = (float) MathHelper.c(entityliving.locZ - EntityEvoker.this.locZ, entityliving.locX - EntityEvoker.this.locX);
             int i;
 
-            if (EntityEvoker.this.h(entityliving) < 9.0D) {
+            if (EntityEvoker.this.h((Entity) entityliving) < 9.0D) {
                 float f1;
 
                 for (i = 0; i < 5; ++i) {
@@ -276,7 +328,7 @@
                 if (!EntityEvoker.this.world.d(blockposition, true) && EntityEvoker.this.world.d(blockposition.down(), true)) {
                     if (!EntityEvoker.this.world.isEmpty(blockposition)) {
                         IBlockData iblockdata = EntityEvoker.this.world.getType(blockposition);
-                        AxisAlignedBB axisalignedbb = iblockdata.d(EntityEvoker.this.world, blockposition);
+                        AxisAlignedBB axisalignedbb = iblockdata.c(EntityEvoker.this.world, blockposition);
 
                         if (axisalignedbb != null) {
                             d4 = axisalignedbb.e;
@@ -299,11 +351,11 @@
         }
 
         protected SoundEffect k() {
-            return SoundEffects.bw;
+            return SoundEffects.bq;
         }
 
-        protected EntityIllagerWizard.Spell l() {
-            return EntityIllagerWizard.Spell.FANGS;
+        protected int l() {
+            return 2;
         }
 
         a(Object object) {
@@ -311,23 +363,85 @@
         }
     }
 
-    class b extends EntityIllagerWizard.b {
+    abstract class d extends PathfinderGoal {
+
+        protected int b;
+        protected int c;
+
+        private d() {}
 
-        private b() {
-            super();
+        public boolean a() {
+            return EntityEvoker.this.getGoalTarget() == null ? false : (EntityEvoker.this.o() ? false : EntityEvoker.this.ticksLived >= this.c);
+        }
+
+        public boolean b() {
+            return EntityEvoker.this.getGoalTarget() != null && this.b > 0;
+        }
+
+        public void c() {
+            this.b = this.m();
+            EntityEvoker.this.b = this.f();
+            this.c = EntityEvoker.this.ticksLived + this.i();
+            EntityEvoker.this.a(this.k(), 1.0F, 1.0F);
+            EntityEvoker.this.c = this.l();
         }
 
         public void e() {
-            if (EntityEvoker.this.getGoalTarget() != null) {
-                EntityEvoker.this.getControllerLook().a(EntityEvoker.this.getGoalTarget(), (float) EntityEvoker.this.O(), (float) EntityEvoker.this.N());
-            } else if (EntityEvoker.this.do() != null) {
-                EntityEvoker.this.getControllerLook().a(EntityEvoker.this.do(), (float) EntityEvoker.this.O(), (float) EntityEvoker.this.N());
+            --this.b;
+            if (this.b == 0) {
+                this.j();
+                EntityEvoker.this.a(SoundEffects.bn, 1.0F, 1.0F);
             }
 
         }
 
-        b(Object object) {
+        protected abstract void j();
+
+        protected int m() {
+            return 20;
+        }
+
+        protected abstract int f();
+
+        protected abstract int i();
+
+        protected abstract SoundEffect k();
+
+        protected abstract int l();
+
+        d(Object object) {
             this();
         }
     }
+
+    class b extends PathfinderGoal {
+
+        public b() {
+            this.a(3);
+        }
+
+        public boolean a() {
+            return EntityEvoker.this.di() > 0;
+        }
+
+        public void c() {
+            super.c();
+            EntityEvoker.this.a(EntityEvoker.this.c);
+            EntityEvoker.this.navigation.o();
+        }
+
+        public void d() {
+            super.d();
+            EntityEvoker.this.a(0);
+        }
+
+        public void e() {
+            if (EntityEvoker.this.getGoalTarget() != null) {
+                EntityEvoker.this.getControllerLook().a(EntityEvoker.this.getGoalTarget(), (float) EntityEvoker.this.cL(), (float) EntityEvoker.this.N());
+            } else if (EntityEvoker.this.dj() != null) {
+                EntityEvoker.this.getControllerLook().a(EntityEvoker.this.dj(), (float) EntityEvoker.this.cL(), (float) EntityEvoker.this.N());
+            }
+
+        }
+    }
 }
