--- a/net/minecraft/server/EnumProtocol.java
+++ b/net/minecraft/server/EnumProtocol.java
@@ -159,7 +159,7 @@
     }
 
     protected EnumProtocol a(EnumProtocolDirection enumprotocoldirection, Class<? extends Packet<?>> oclass) {
-        Object object = (BiMap) this.h.get(enumprotocoldirection);
+        BiMap<Integer, Class<? extends Packet<?>>> object = (BiMap) this.h.get(enumprotocoldirection); // SportBukkit - fix decomp error
 
         if (object == null) {
             object = HashBiMap.create();
