--- a/net/minecraft/server/IInventory.java
+++ b/net/minecraft/server/IInventory.java
@@ -1,10 +1,12 @@
 package net.minecraft.server;
 
+import org.bukkit.craftbukkit.entity.CraftHumanEntity; // CraftBukkit
+
 public interface IInventory extends INamableTileEntity {
 
     int getSize();
 
-    boolean x_();
+    boolean w_();
 
     ItemStack getItem(int i);
 
@@ -33,4 +35,22 @@
     int h();
 
     void clear();
+
+    // CraftBukkit start
+    java.util.List<ItemStack> getContents();
+
+    void onOpen(CraftHumanEntity who);
+
+    void onClose(CraftHumanEntity who);
+
+    java.util.List<org.bukkit.entity.HumanEntity> getViewers();
+
+    org.bukkit.inventory.InventoryHolder getOwner();
+
+    void setMaxStackSize(int size);
+
+    org.bukkit.Location getLocation();
+
+    int MAX_STACK = 64;
+    // CraftBukkit end
 }
