--- a/net/minecraft/server/ItemSkull.java
+++ b/net/minecraft/server/ItemSkull.java
@@ -83,15 +83,6 @@
         }
     }
 
-    public void a(CreativeModeTab creativemodetab, NonNullList<ItemStack> nonnulllist) {
-        if (this.a(creativemodetab)) {
-            for (int i = 0; i < ItemSkull.a.length; ++i) {
-                nonnulllist.add(new ItemStack(this, 1, i));
-            }
-        }
-
-    }
-
     public int filterData(int i) {
         return i;
     }
@@ -133,6 +124,15 @@
             nbttagcompound.set("SkullOwner", GameProfileSerializer.serialize(new NBTTagCompound(), gameprofile));
             return true;
         } else {
+            // CraftBukkit start
+            NBTTagList textures = nbttagcompound.getCompound("SkullOwner").getCompound("Properties").getList("textures", 10); // Safe due to method contracts
+            for (int i = 0; i < textures.size(); i++) {
+                if (textures.get(i) instanceof NBTTagCompound && !((NBTTagCompound) textures.get(i)).hasKeyOfType("Signature", 8) && ((NBTTagCompound) textures.get(i)).getString("Value").trim().isEmpty()) {
+                    nbttagcompound.remove("SkullOwner");
+                    break;
+                }
+            }
+            // CraftBukkit end
             return false;
         }
     }
