--- a/net/minecraft/server/ItemReed.java
+++ b/net/minecraft/server/ItemReed.java
@@ -2,7 +2,7 @@
 
 public class ItemReed extends Item {
 
-    private final Block a;
+    public final Block a; // PAIL: private->public
 
     public ItemReed(Block block) {
         this.a = block;
@@ -30,9 +30,6 @@
                 if (iblockdata1.getBlock() == this.a) {
                     ItemBlock.a(world, entityhuman, blockposition, itemstack);
                     iblockdata1.getBlock().postPlace(world, blockposition, iblockdata1, entityhuman, itemstack);
-                    if (entityhuman instanceof EntityPlayer) {
-                        CriterionTriggers.x.a((EntityPlayer) entityhuman, blockposition, itemstack);
-                    }
                 }
 
                 SoundEffectType soundeffecttype = this.a.getStepSound();
