--- a/net/minecraft/server/PacketEncoder.java
+++ b/net/minecraft/server/PacketEncoder.java
@@ -41,14 +41,14 @@
                 try {
                     packet.b(packetdataserializer);
                 } catch (Throwable throwable) {
-                    PacketEncoder.a.error(throwable);
+                PacketEncoder.a.error("Exception serializing packet " + packet, throwable); // SportBukkit - better error message
                 }
 
             }
         }
     }
 
-    protected void encode(ChannelHandlerContext channelhandlercontext, Object object, ByteBuf bytebuf) throws Exception {
+    protected void encode(ChannelHandlerContext channelhandlercontext, Packet<?> object, ByteBuf bytebuf) throws Exception { // SportBukkit - fix decomp error
         this.a(channelhandlercontext, (Packet) object, bytebuf);
     }
 }
