--- a/net/minecraft/server/PacketPlayOutPlayerInfo.java
+++ b/net/minecraft/server/PacketPlayOutPlayerInfo.java
@@ -11,12 +11,27 @@
 
 public class PacketPlayOutPlayerInfo implements Packet<PacketListenerPlayOut> {
 
-    private PacketPlayOutPlayerInfo.EnumPlayerInfoAction a;
-    private final List<PacketPlayOutPlayerInfo.PlayerInfoData> b = Lists.newArrayList();
+    private EnumPlayerInfoAction a;
+    private final List<PlayerInfoData> b = Lists.newArrayList();
+
+    // SportBukkit start - add constructor
+    public PacketPlayOutPlayerInfo(EnumPlayerInfoAction a) {
+        this.a = a;
+    }
+
+    public void add(PacketPlayOutPlayerInfo.PlayerInfoData info) {
+        this.b.add(info);
+    }
+
+    public boolean isEmpty() {
+        return this.b.isEmpty();
+    }
+    // SportBukkit end
+
 
     public PacketPlayOutPlayerInfo() {}
 
-    public PacketPlayOutPlayerInfo(PacketPlayOutPlayerInfo.EnumPlayerInfoAction packetplayoutplayerinfo_enumplayerinfoaction, EntityPlayer... aentityplayer) {
+    public PacketPlayOutPlayerInfo(EnumPlayerInfoAction packetplayoutplayerinfo_enumplayerinfoaction, EntityPlayer... aentityplayer) {
         this.a = packetplayoutplayerinfo_enumplayerinfoaction;
         EntityPlayer[] aentityplayer1 = aentityplayer;
         int i = aentityplayer.length;
@@ -24,25 +39,25 @@
         for (int j = 0; j < i; ++j) {
             EntityPlayer entityplayer = aentityplayer1[j];
 
-            this.b.add(new PacketPlayOutPlayerInfo.PlayerInfoData(entityplayer.getProfile(), entityplayer.ping, entityplayer.playerInteractManager.getGameMode(), entityplayer.getPlayerListName()));
+            this.b.add(new PlayerInfoData(entityplayer.getProfile(), entityplayer.ping, entityplayer.playerInteractManager.getGameMode(), entityplayer.getPlayerListName()));
         }
 
     }
 
-    public PacketPlayOutPlayerInfo(PacketPlayOutPlayerInfo.EnumPlayerInfoAction packetplayoutplayerinfo_enumplayerinfoaction, Iterable<EntityPlayer> iterable) {
+    public PacketPlayOutPlayerInfo(EnumPlayerInfoAction packetplayoutplayerinfo_enumplayerinfoaction, Iterable<EntityPlayer> iterable) {
         this.a = packetplayoutplayerinfo_enumplayerinfoaction;
         Iterator iterator = iterable.iterator();
 
         while (iterator.hasNext()) {
             EntityPlayer entityplayer = (EntityPlayer) iterator.next();
 
-            this.b.add(new PacketPlayOutPlayerInfo.PlayerInfoData(entityplayer.getProfile(), entityplayer.ping, entityplayer.playerInteractManager.getGameMode(), entityplayer.getPlayerListName()));
+            this.b.add(new PlayerInfoData(entityplayer.getProfile(), entityplayer.ping, entityplayer.playerInteractManager.getGameMode(), entityplayer.getPlayerListName()));
         }
 
     }
 
     public void a(PacketDataSerializer packetdataserializer) throws IOException {
-        this.a = (PacketPlayOutPlayerInfo.EnumPlayerInfoAction) packetdataserializer.a(PacketPlayOutPlayerInfo.EnumPlayerInfoAction.class);
+        this.a = (EnumPlayerInfoAction) packetdataserializer.a(EnumPlayerInfoAction.class);
         int i = packetdataserializer.g();
 
         for (int j = 0; j < i; ++j) {
@@ -95,7 +110,7 @@
                 gameprofile = new GameProfile(packetdataserializer.i(), (String) null);
             }
 
-            this.b.add(new PacketPlayOutPlayerInfo.PlayerInfoData(gameprofile, k, enumgamemode, ichatbasecomponent));
+            this.b.add(new PlayerInfoData(gameprofile, k, enumgamemode, ichatbasecomponent));
         }
 
     }
@@ -106,7 +121,7 @@
         Iterator iterator = this.b.iterator();
 
         while (iterator.hasNext()) {
-            PacketPlayOutPlayerInfo.PlayerInfoData packetplayoutplayerinfo_playerinfodata = (PacketPlayOutPlayerInfo.PlayerInfoData) iterator.next();
+            PlayerInfoData packetplayoutplayerinfo_playerinfodata = (PlayerInfoData) iterator.next();
 
             switch (this.a) {
             case ADD_PLAYER:
@@ -130,6 +145,15 @@
 
                 packetdataserializer.d(packetplayoutplayerinfo_playerinfodata.c().getId());
                 packetdataserializer.d(packetplayoutplayerinfo_playerinfodata.b());
+
+                // SportBukkit start
+                if (packetplayoutplayerinfo_playerinfodata.displayName != null) {
+                    packetdataserializer.writeBoolean(true);
+                    packetdataserializer.a(net.md_5.bungee.chat.ComponentSerializer.toString(packetplayoutplayerinfo_playerinfodata.displayName));
+                    break;
+                }
+                // SportBukkit end
+
                 if (packetplayoutplayerinfo_playerinfodata.d() == null) {
                     packetdataserializer.writeBoolean(false);
                 } else {
@@ -150,6 +174,15 @@
 
             case UPDATE_DISPLAY_NAME:
                 packetdataserializer.a(packetplayoutplayerinfo_playerinfodata.a().getId());
+
+                // SportBukkit start
+                if (packetplayoutplayerinfo_playerinfodata.displayName != null) {
+                    packetdataserializer.writeBoolean(true);
+                    packetdataserializer.a(net.md_5.bungee.chat.ComponentSerializer.toString(packetplayoutplayerinfo_playerinfodata.displayName));
+                    break;
+                }
+                // SportBukkit end
+
                 if (packetplayoutplayerinfo_playerinfodata.d() == null) {
                     packetdataserializer.writeBoolean(false);
                 } else {
@@ -175,6 +208,8 @@
 
     public class PlayerInfoData {
 
+        public net.md_5.bungee.api.chat.BaseComponent[] displayName; // SportBukkit
+
         private final int b;
         private final EnumGamemode c;
         private final GameProfile d;
