--- a/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java
+++ b/net/minecraft/server/PacketPlayOutSpawnEntityLiving.java
@@ -21,12 +21,30 @@
     private DataWatcher m;
     private List<DataWatcher.Item<?>> n;
 
+    // SportBukkit start - add constructor
+    public PacketPlayOutSpawnEntityLiving(int id, UUID uuid, int type, double xPos, double yPos, double zPos, int xVel, int yVel, int zVel, byte yaw, byte pitch, byte headPitch, List<DataWatcher.Item<?>> metadata) {
+        this.a = id;
+        this.b = uuid;
+        this.c = type;
+        this.d = xPos;
+        this.e = yPos;
+        this.f = zPos;
+        this.g = xVel;
+        this.h = yVel;
+        this.i = zVel;
+        this.j = yaw;
+        this.k = pitch;
+        this.l = headPitch;
+        this.n = metadata;
+    }
+    // SportBukkit end
+
     public PacketPlayOutSpawnEntityLiving() {}
 
     public PacketPlayOutSpawnEntityLiving(EntityLiving entityliving) {
         this.a = entityliving.getId();
         this.b = entityliving.getUniqueID();
-        this.c = EntityTypes.b.a((Object) entityliving.getClass());
+        this.c = EntityTypes.b.a(entityliving.getClass()); // SportBukkit - fix decomp
         this.d = entityliving.locX;
         this.e = entityliving.locY;
         this.f = entityliving.locZ;
@@ -66,6 +84,11 @@
         this.h = (int) (d2 * 8000.0D);
         this.i = (int) (d3 * 8000.0D);
         this.m = entityliving.getDataWatcher();
+
+        // SportBukkit start - send unshared copy of metadata
+        this.n = this.m.c();
+        DataWatcher.deepCopy(this.n);
+        // SportBukkit end
     }
 
     public void a(PacketDataSerializer packetdataserializer) throws IOException {
@@ -97,7 +120,11 @@
         packetdataserializer.writeShort(this.g);
         packetdataserializer.writeShort(this.h);
         packetdataserializer.writeShort(this.i);
-        this.m.a(packetdataserializer);
+
+        // SportBukkit start - send unshared copy of metadata
+        // this.m.a(packetdataserializer);
+        DataWatcher.a(this.n, packetdataserializer);
+        // SportBukkit end
     }
 
     public void a(PacketListenerPlayOut packetlistenerplayout) {
