--- a/net/minecraft/server/PacketPlayInResourcePackStatus.java
+++ b/net/minecraft/server/PacketPlayInResourcePackStatus.java
@@ -4,16 +4,16 @@
 
 public class PacketPlayInResourcePackStatus implements Packet<PacketListenerPlayIn> {
 
-    public PacketPlayInResourcePackStatus.EnumResourcePackStatus status;
+    public EnumResourcePackStatus status;
 
     public PacketPlayInResourcePackStatus() {}
 
-    public PacketPlayInResourcePackStatus(PacketPlayInResourcePackStatus.EnumResourcePackStatus packetplayinresourcepackstatus_enumresourcepackstatus) {
+    public PacketPlayInResourcePackStatus(EnumResourcePackStatus packetplayinresourcepackstatus_enumresourcepackstatus) {
         this.status = packetplayinresourcepackstatus_enumresourcepackstatus;
     }
 
     public void a(PacketDataSerializer packetdataserializer) throws IOException {
-        this.status = (PacketPlayInResourcePackStatus.EnumResourcePackStatus) packetdataserializer.a(PacketPlayInResourcePackStatus.EnumResourcePackStatus.class);
+        this.status = (EnumResourcePackStatus) packetdataserializer.a(EnumResourcePackStatus.class);
     }
 
     public void b(PacketDataSerializer packetdataserializer) throws IOException {
