--- a/net/minecraft/server/PacketDataSerializer.java
+++ b/net/minecraft/server/PacketDataSerializer.java
@@ -1,12 +1,13 @@
 package net.minecraft.server;
 
+import com.google.common.base.Charsets;
 import io.netty.buffer.ByteBuf;
 import io.netty.buffer.ByteBufAllocator;
 import io.netty.buffer.ByteBufInputStream;
 import io.netty.buffer.ByteBufOutputStream;
+import io.netty.buffer.ByteBufProcessor;
 import io.netty.handler.codec.DecoderException;
 import io.netty.handler.codec.EncoderException;
-import io.netty.util.ByteProcessor;
 import java.io.DataInput;
 import java.io.DataOutput;
 import java.io.IOException;
@@ -14,17 +15,19 @@
 import java.io.OutputStream;
 import java.nio.ByteBuffer;
 import java.nio.ByteOrder;
-import java.nio.channels.FileChannel;
 import java.nio.channels.GatheringByteChannel;
 import java.nio.channels.ScatteringByteChannel;
 import java.nio.charset.Charset;
-import java.nio.charset.StandardCharsets;
-import java.util.Date;
 import java.util.UUID;
 import javax.annotation.Nullable;
 
+import org.bukkit.craftbukkit.inventory.CraftItemStack; // CraftBukkit
+import org.bukkit.craftbukkit.protocol.Protocol; // SportBukkit
+
 public class PacketDataSerializer extends ByteBuf {
 
+    public int protocolVersion = Protocol.LATEST; // SportBukkit
+
     private final ByteBuf a;
 
     public PacketDataSerializer(ByteBuf bytebuf) {
@@ -116,6 +119,7 @@
         return BlockPosition.fromLong(this.readLong());
     }
 
+    public PacketDataSerializer writeBlockPosition(BlockPosition pos) { return a(pos); } // SportBukkit - alias
     public PacketDataSerializer a(BlockPosition blockposition) {
         this.writeLong(blockposition.asLong());
         return this;
@@ -125,18 +129,21 @@
         return IChatBaseComponent.ChatSerializer.a(this.e(32767));
     }
 
+    public PacketDataSerializer writeChatComponent(IChatBaseComponent chat) { return a(chat); } // SportBukkit - alias
     public PacketDataSerializer a(IChatBaseComponent ichatbasecomponent) {
         return this.a(IChatBaseComponent.ChatSerializer.a(ichatbasecomponent));
     }
 
     public <T extends Enum<T>> T a(Class<T> oclass) {
-        return ((Enum[]) oclass.getEnumConstants())[this.g()];
+        return ((T[]) oclass.getEnumConstants())[this.g()]; // CraftBukkit - fix decompile error
     }
 
+    public PacketDataSerializer writeOrdinal(Enum<?> e) { return a(e); } // SportBukkit - alias
     public PacketDataSerializer a(Enum<?> oenum) {
         return this.d(oenum.ordinal());
     }
 
+    public int readVarInt() { return g(); } // SportBukkit - alias
     public int g() {
         int i = 0;
         int j = 0;
@@ -181,6 +188,7 @@
         return new UUID(this.readLong(), this.readLong());
     }
 
+    public PacketDataSerializer writeVarInt(int i) { return d(i); } // SportBukkit - alias
     public PacketDataSerializer d(int i) {
         while ((i & -128) != 0) {
             this.writeByte(i & 127 | 128);
@@ -207,7 +215,7 @@
         } else {
             try {
                 NBTCompressedStreamTools.a(nbttagcompound, (DataOutput) (new ByteBufOutputStream(this)));
-            } catch (IOException ioexception) {
+            } catch (Exception ioexception) { // CraftBukkit - IOException -> Exception
                 throw new EncoderException(ioexception);
             }
         }
@@ -234,15 +242,19 @@
     }
 
     public PacketDataSerializer a(ItemStack itemstack) {
-        if (itemstack.isEmpty()) {
+        if (itemstack.isEmpty() || itemstack.getItem() == null) { // CraftBukkit - NPE fix itemstack.getItem()
             this.writeShort(-1);
+        // SportBukkit start - legacy clients will crash if they get an unknown item
+        } else if(protocolVersion < itemstack.getItem().proto()) {
+            this.writeShort(-1);
+        // SportBukkit end
         } else {
             this.writeShort(Item.getId(itemstack.getItem()));
             this.writeByte(itemstack.getCount());
             this.writeShort(itemstack.getData());
             NBTTagCompound nbttagcompound = null;
 
-            if (itemstack.getItem().usesDurability() || itemstack.getItem().p()) {
+            if (itemstack.getItem().usesDurability() || itemstack.getItem().q()) {
                 nbttagcompound = itemstack.getTag();
             }
 
@@ -263,6 +275,11 @@
             ItemStack itemstack = new ItemStack(Item.getById(short0), b0, short1);
 
             itemstack.setTag(this.j());
+            // CraftBukkit start
+            if (itemstack.getTag() != null) {
+                CraftItemStack.setItemMeta(itemstack, CraftItemStack.getItemMeta(itemstack));
+            }
+            // CraftBukkit end
             return itemstack;
         }
     }
@@ -275,9 +292,8 @@
         } else if (j < 0) {
             throw new DecoderException("The received encoded string buffer length is less than zero! Weird string!");
         } else {
-            String s = this.toString(this.readerIndex(), j, StandardCharsets.UTF_8);
+            String s = new String(this.readBytes(j).array(), Charsets.UTF_8);
 
-            this.readerIndex(this.readerIndex() + j);
             if (s.length() > i) {
                 throw new DecoderException("The received string length is longer than maximum allowed (" + j + " > " + i + ")");
             } else {
@@ -287,7 +303,7 @@
     }
 
     public PacketDataSerializer a(String s) {
-        byte[] abyte = s.getBytes(StandardCharsets.UTF_8);
+        byte[] abyte = s.getBytes(Charsets.UTF_8);
 
         if (abyte.length > 32767) {
             throw new EncoderException("String too big (was " + abyte.length + " bytes encoded, max " + 32767 + ")");
@@ -298,24 +314,6 @@
         }
     }
 
-    public MinecraftKey l() {
-        return new MinecraftKey(this.e(32767));
-    }
-
-    public PacketDataSerializer a(MinecraftKey minecraftkey) {
-        this.a(minecraftkey.toString());
-        return this;
-    }
-
-    public Date m() {
-        return new Date(this.readLong());
-    }
-
-    public PacketDataSerializer a(Date date) {
-        this.writeLong(date.getTime());
-        return this;
-    }
-
     public int capacity() {
         return this.a.capacity();
     }
@@ -348,14 +346,6 @@
         return this.a.isDirect();
     }
 
-    public boolean isReadOnly() {
-        return this.a.isReadOnly();
-    }
-
-    public ByteBuf asReadOnly() {
-        return this.a.asReadOnly();
-    }
-
     public int readerIndex() {
         return this.a.readerIndex();
     }
@@ -456,58 +446,30 @@
         return this.a.getShort(i);
     }
 
-    public short getShortLE(int i) {
-        return this.a.getShortLE(i);
-    }
-
     public int getUnsignedShort(int i) {
         return this.a.getUnsignedShort(i);
     }
 
-    public int getUnsignedShortLE(int i) {
-        return this.a.getUnsignedShortLE(i);
-    }
-
     public int getMedium(int i) {
         return this.a.getMedium(i);
     }
 
-    public int getMediumLE(int i) {
-        return this.a.getMediumLE(i);
-    }
-
     public int getUnsignedMedium(int i) {
         return this.a.getUnsignedMedium(i);
     }
 
-    public int getUnsignedMediumLE(int i) {
-        return this.a.getUnsignedMediumLE(i);
-    }
-
     public int getInt(int i) {
         return this.a.getInt(i);
     }
 
-    public int getIntLE(int i) {
-        return this.a.getIntLE(i);
-    }
-
     public long getUnsignedInt(int i) {
         return this.a.getUnsignedInt(i);
     }
 
-    public long getUnsignedIntLE(int i) {
-        return this.a.getUnsignedIntLE(i);
-    }
-
     public long getLong(int i) {
         return this.a.getLong(i);
     }
 
-    public long getLongLE(int i) {
-        return this.a.getLongLE(i);
-    }
-
     public char getChar(int i) {
         return this.a.getChar(i);
     }
@@ -552,14 +514,6 @@
         return this.a.getBytes(i, gatheringbytechannel, j);
     }
 
-    public int getBytes(int i, FileChannel filechannel, long j, int k) throws IOException {
-        return this.a.getBytes(i, filechannel, j, k);
-    }
-
-    public CharSequence getCharSequence(int i, int j, Charset charset) {
-        return this.a.getCharSequence(i, j, charset);
-    }
-
     public ByteBuf setBoolean(int i, boolean flag) {
         return this.a.setBoolean(i, flag);
     }
@@ -572,34 +526,18 @@
         return this.a.setShort(i, j);
     }
 
-    public ByteBuf setShortLE(int i, int j) {
-        return this.a.setShortLE(i, j);
-    }
-
     public ByteBuf setMedium(int i, int j) {
         return this.a.setMedium(i, j);
     }
 
-    public ByteBuf setMediumLE(int i, int j) {
-        return this.a.setMediumLE(i, j);
-    }
-
     public ByteBuf setInt(int i, int j) {
         return this.a.setInt(i, j);
     }
 
-    public ByteBuf setIntLE(int i, int j) {
-        return this.a.setIntLE(i, j);
-    }
-
     public ByteBuf setLong(int i, long j) {
         return this.a.setLong(i, j);
     }
 
-    public ByteBuf setLongLE(int i, long j) {
-        return this.a.setLongLE(i, j);
-    }
-
     public ByteBuf setChar(int i, int j) {
         return this.a.setChar(i, j);
     }
@@ -644,18 +582,10 @@
         return this.a.setBytes(i, scatteringbytechannel, j);
     }
 
-    public int setBytes(int i, FileChannel filechannel, long j, int k) throws IOException {
-        return this.a.setBytes(i, filechannel, j, k);
-    }
-
     public ByteBuf setZero(int i, int j) {
         return this.a.setZero(i, j);
     }
 
-    public int setCharSequence(int i, CharSequence charsequence, Charset charset) {
-        return this.a.setCharSequence(i, charsequence, charset);
-    }
-
     public boolean readBoolean() {
         return this.a.readBoolean();
     }
@@ -672,58 +602,30 @@
         return this.a.readShort();
     }
 
-    public short readShortLE() {
-        return this.a.readShortLE();
-    }
-
     public int readUnsignedShort() {
         return this.a.readUnsignedShort();
     }
 
-    public int readUnsignedShortLE() {
-        return this.a.readUnsignedShortLE();
-    }
-
     public int readMedium() {
         return this.a.readMedium();
     }
 
-    public int readMediumLE() {
-        return this.a.readMediumLE();
-    }
-
     public int readUnsignedMedium() {
         return this.a.readUnsignedMedium();
     }
 
-    public int readUnsignedMediumLE() {
-        return this.a.readUnsignedMediumLE();
-    }
-
     public int readInt() {
         return this.a.readInt();
     }
 
-    public int readIntLE() {
-        return this.a.readIntLE();
-    }
-
     public long readUnsignedInt() {
         return this.a.readUnsignedInt();
     }
 
-    public long readUnsignedIntLE() {
-        return this.a.readUnsignedIntLE();
-    }
-
     public long readLong() {
         return this.a.readLong();
     }
 
-    public long readLongLE() {
-        return this.a.readLongLE();
-    }
-
     public char readChar() {
         return this.a.readChar();
     }
@@ -744,10 +646,6 @@
         return this.a.readSlice(i);
     }
 
-    public ByteBuf readRetainedSlice(int i) {
-        return this.a.readRetainedSlice(i);
-    }
-
     public ByteBuf readBytes(ByteBuf bytebuf) {
         return this.a.readBytes(bytebuf);
     }
@@ -780,14 +678,6 @@
         return this.a.readBytes(gatheringbytechannel, i);
     }
 
-    public CharSequence readCharSequence(int i, Charset charset) {
-        return this.a.readCharSequence(i, charset);
-    }
-
-    public int readBytes(FileChannel filechannel, long i, int j) throws IOException {
-        return this.a.readBytes(filechannel, i, j);
-    }
-
     public ByteBuf skipBytes(int i) {
         return this.a.skipBytes(i);
     }
@@ -804,34 +694,18 @@
         return this.a.writeShort(i);
     }
 
-    public ByteBuf writeShortLE(int i) {
-        return this.a.writeShortLE(i);
-    }
-
     public ByteBuf writeMedium(int i) {
         return this.a.writeMedium(i);
     }
 
-    public ByteBuf writeMediumLE(int i) {
-        return this.a.writeMediumLE(i);
-    }
-
     public ByteBuf writeInt(int i) {
         return this.a.writeInt(i);
     }
 
-    public ByteBuf writeIntLE(int i) {
-        return this.a.writeIntLE(i);
-    }
-
     public ByteBuf writeLong(long i) {
         return this.a.writeLong(i);
     }
 
-    public ByteBuf writeLongLE(long i) {
-        return this.a.writeLongLE(i);
-    }
-
     public ByteBuf writeChar(int i) {
         return this.a.writeChar(i);
     }
@@ -876,18 +750,10 @@
         return this.a.writeBytes(scatteringbytechannel, i);
     }
 
-    public int writeBytes(FileChannel filechannel, long i, int j) throws IOException {
-        return this.a.writeBytes(filechannel, i, j);
-    }
-
     public ByteBuf writeZero(int i) {
         return this.a.writeZero(i);
     }
 
-    public int writeCharSequence(CharSequence charsequence, Charset charset) {
-        return this.a.writeCharSequence(charsequence, charset);
-    }
-
     public int indexOf(int i, int j, byte b0) {
         return this.a.indexOf(i, j, b0);
     }
@@ -904,20 +770,20 @@
         return this.a.bytesBefore(i, j, b0);
     }
 
-    public int forEachByte(ByteProcessor byteprocessor) {
-        return this.a.forEachByte(byteprocessor);
+    public int forEachByte(ByteBufProcessor bytebufprocessor) {
+        return this.a.forEachByte(bytebufprocessor);
     }
 
-    public int forEachByte(int i, int j, ByteProcessor byteprocessor) {
-        return this.a.forEachByte(i, j, byteprocessor);
+    public int forEachByte(int i, int j, ByteBufProcessor bytebufprocessor) {
+        return this.a.forEachByte(i, j, bytebufprocessor);
     }
 
-    public int forEachByteDesc(ByteProcessor byteprocessor) {
-        return this.a.forEachByteDesc(byteprocessor);
+    public int forEachByteDesc(ByteBufProcessor bytebufprocessor) {
+        return this.a.forEachByteDesc(bytebufprocessor);
     }
 
-    public int forEachByteDesc(int i, int j, ByteProcessor byteprocessor) {
-        return this.a.forEachByteDesc(i, j, byteprocessor);
+    public int forEachByteDesc(int i, int j, ByteBufProcessor bytebufprocessor) {
+        return this.a.forEachByteDesc(i, j, bytebufprocessor);
     }
 
     public ByteBuf copy() {
@@ -932,26 +798,14 @@
         return this.a.slice();
     }
 
-    public ByteBuf retainedSlice() {
-        return this.a.retainedSlice();
-    }
-
     public ByteBuf slice(int i, int j) {
         return this.a.slice(i, j);
     }
 
-    public ByteBuf retainedSlice(int i, int j) {
-        return this.a.retainedSlice(i, j);
-    }
-
     public ByteBuf duplicate() {
         return this.a.duplicate();
     }
 
-    public ByteBuf retainedDuplicate() {
-        return this.a.retainedDuplicate();
-    }
-
     public int nioBufferCount() {
         return this.a.nioBufferCount();
     }
@@ -1028,14 +882,6 @@
         return this.a.retain();
     }
 
-    public ByteBuf touch() {
-        return this.a.touch();
-    }
-
-    public ByteBuf touch(Object object) {
-        return this.a.touch(object);
-    }
-
     public int refCnt() {
         return this.a.refCnt();
     }
