--- a/net/minecraft/server/PathfinderGoalDefendVillage.java
+++ b/net/minecraft/server/PathfinderGoalDefendVillage.java
@@ -12,7 +12,7 @@
     }
 
     public boolean a() {
-        Village village = this.a.p();
+        Village village = this.a.o();
 
         if (village == null) {
             return false;
@@ -32,7 +32,7 @@
     }
 
     public void c() {
-        this.a.setGoalTarget(this.b);
+        this.a.setGoalTarget(this.b, org.bukkit.event.entity.EntityTargetEvent.TargetReason.DEFEND_VILLAGE, true); // CraftBukkit - reason
         super.c();
     }
 }
