--- a/net/minecraft/server/PathfinderGoalOwnerHurtTarget.java
+++ b/net/minecraft/server/PathfinderGoalOwnerHurtTarget.java
@@ -21,8 +21,8 @@
             if (entityliving == null) {
                 return false;
             } else {
-                this.b = entityliving.bS();
-                int i = entityliving.bT();
+                this.b = entityliving.bM();
+                int i = entityliving.bN();
 
                 return i != this.c && this.a(this.b, false) && this.a.a(this.b, entityliving);
             }
@@ -30,11 +30,11 @@
     }
 
     public void c() {
-        this.e.setGoalTarget(this.b);
+        this.e.setGoalTarget(this.b, org.bukkit.event.entity.EntityTargetEvent.TargetReason.OWNER_ATTACKED_TARGET, true); // CraftBukkit - reason
         EntityLiving entityliving = this.a.getOwner();
 
         if (entityliving != null) {
-            this.c = entityliving.bT();
+            this.c = entityliving.bN();
         }
 
         super.c();
