--- a/net/minecraft/server/PotionUtil.java
+++ b/net/minecraft/server/PotionUtil.java
@@ -1,6 +1,6 @@
 package net.minecraft.server;
 
-import com.google.common.base.MoreObjects;
+import com.google.common.base.Objects;
 import com.google.common.collect.Lists;
 import java.util.ArrayList;
 import java.util.Collection;
@@ -137,7 +137,7 @@
         if (collection.isEmpty()) {
             return itemstack;
         } else {
-            NBTTagCompound nbttagcompound = (NBTTagCompound) MoreObjects.firstNonNull(itemstack.getTag(), new NBTTagCompound());
+            NBTTagCompound nbttagcompound = (NBTTagCompound) Objects.firstNonNull(itemstack.getTag(), new NBTTagCompound());
             NBTTagList nbttaglist = nbttagcompound.getList("CustomPotionEffects", 9);
             Iterator iterator = collection.iterator();
 
