--- a/net/minecraft/server/SoundEffectType.java
+++ b/net/minecraft/server/SoundEffectType.java
@@ -40,15 +40,23 @@
         return this.n;
     }
 
+    public SoundEffect stepSound() { return this.d(); } // SportBukkit - alias for following method
     public SoundEffect d() {
         return this.p;
     }
 
+    public SoundEffect placeSound() { return this.e(); } // SportBukkit - alias for following method
     public SoundEffect e() {
         return this.q;
     }
 
+    public SoundEffect fallSound() { return this.g(); } // SportBukkit - alias for following method
     public SoundEffect g() {
         return this.s;
     }
+
+    // SportBukkit start - add missing accessors
+    public SoundEffect breakSound() { return this.o;}
+    public SoundEffect hitSound() { return this.r;}
+    // SportBukkit end
 }
