--- a/net/minecraft/server/SoundEffectType.java
+++ b/net/minecraft/server/SoundEffectType.java
@@ -2,7 +2,7 @@
 
 public class SoundEffectType {
 
-    public static final SoundEffectType a = new SoundEffectType(1.0F, 1.0F, SoundEffects.iZ, SoundEffects.jh, SoundEffects.je, SoundEffects.jd, SoundEffects.jc);
+    public static final SoundEffectType a = new SoundEffectType(1.0F, 1.0F, SoundEffects.iW, SoundEffects.je, SoundEffects.jb, SoundEffects.ja, SoundEffects.iZ);
     public static final SoundEffectType b = new SoundEffectType(1.0F, 1.0F, SoundEffects.cr, SoundEffects.cv, SoundEffects.cu, SoundEffects.ct, SoundEffects.cs);
     public static final SoundEffectType c = new SoundEffectType(1.0F, 1.0F, SoundEffects.cm, SoundEffects.cq, SoundEffects.cp, SoundEffects.co, SoundEffects.cn);
     public static final SoundEffectType d = new SoundEffectType(1.0F, 1.0F, SoundEffects.hI, SoundEffects.hQ, SoundEffects.hN, SoundEffects.hM, SoundEffects.hL);
@@ -40,15 +40,23 @@
         return this.n;
     }
 
+    public SoundEffect stepSound() { return this.d(); } // SportBukkit - alias for following method
     public SoundEffect d() {
         return this.p;
     }
 
+    public SoundEffect placeSound() { return this.e(); } // SportBukkit - alias for following method
     public SoundEffect e() {
         return this.q;
     }
 
+    public SoundEffect fallSound() { return this.g(); } // SportBukkit - alias for following method
     public SoundEffect g() {
         return this.s;
     }
+
+    // SportBukkit start - add missing accessors
+    public SoundEffect breakSound() { return this.o;}
+    public SoundEffect hitSound() { return this.r;}
+    // SportBukkit end
 }
