--- a/net/minecraft/server/TileEntitySkull.java
+++ b/net/minecraft/server/TileEntitySkull.java
@@ -41,8 +41,8 @@
         return nbttagcompound;
     }
 
-    public void load(NBTTagCompound nbttagcompound) {
-        super.load(nbttagcompound);
+    public void a(NBTTagCompound nbttagcompound) {
+        super.a(nbttagcompound);
         this.a = nbttagcompound.getByte("SkullType");
         this.rotation = nbttagcompound.getByte("Rot");
         if (this.a == 3) {
@@ -106,6 +106,9 @@
         if (gameprofile != null && !UtilColor.b(gameprofile.getName())) {
             if (gameprofile.isComplete() && gameprofile.getProperties().containsKey("textures")) {
                 return gameprofile;
+            } else if(!org.bukkit.Bukkit.getFetchSkulls()) {
+                org.bukkit.Bukkit.getLogger().warning("Skull-fetching is disabled, so not fetching skin for '" + gameprofile.getName() + "'");
+                return gameprofile;
             } else if (TileEntitySkull.j != null && TileEntitySkull.k != null) {
                 GameProfile gameprofile1 = TileEntitySkull.j.getProfile(gameprofile.getName());
 
