--- a/net/minecraft/server/TileEntity.java
+++ b/net/minecraft/server/TileEntity.java
@@ -4,6 +4,8 @@
 import org.apache.logging.log4j.LogManager;
 import org.apache.logging.log4j.Logger;
 
+import org.bukkit.inventory.InventoryHolder; // CraftBukkit
+
 public abstract class TileEntity {
 
     private static final Logger a = LogManager.getLogger();
@@ -142,6 +144,7 @@
         return this.d;
     }
 
+    public void remove() { z(); } // SportBukkit - alias
     public void z() {
         this.d = true;
     }
@@ -170,7 +173,12 @@
             }
         });
         if (this.world != null) {
-            CrashReportSystemDetails.a(crashreportsystemdetails, this.position, this.getBlock(), this.v());
+            // Paper start - Prevent TileEntity and Entity crashes
+            Block block = this.getBlock();
+            if (block != null) {
+                CrashReportSystemDetails.a(crashreportsystemdetails, this.position, this.getBlock(), this.v());
+            }
+            // Paper end
             crashreportsystemdetails.a("Actual block type", new CrashReportCallable() {
                 public String a() throws Exception {
                     int i = Block.getId(TileEntity.this.world.getType(TileEntity.this.position).getBlock());
@@ -251,4 +259,13 @@
         a("shulker_box", TileEntityShulkerBox.class);
         a("bed", TileEntityBed.class);
     }
+
+    // CraftBukkit start - add method
+    public InventoryHolder getOwner() {
+        if (world == null) return null;
+        org.bukkit.block.BlockState state = world.getWorld().getBlockAt(position.getX(), position.getY(), position.getZ()).getState();
+        if (state instanceof InventoryHolder) return (InventoryHolder) state;
+        return null;
+    }
+    // CraftBukkit end
 }
