From 36f3c92108596991ecaf466d05f4dab9e582920a Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Wed, 4 Feb 2015 03:39:54 -0500
Subject: [PATCH] Add setting to disable skull fetching


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 2a56468e..2610b308 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -314,6 +314,13 @@ public final class Bukkit {
     }
 
     /**
+     * @see Server#getFetchSkulls()
+     */
+    public static boolean getFetchSkulls() {
+        return server.getFetchSkulls();
+    }
+
+    /**
      * Gets default ticks per animal spawns value.
      * <p>
      * <b>Example Usage:</b>
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 1829b47e..35501556 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -256,6 +256,11 @@ public interface Server extends PluginMessageRecipient {
     public long getConnectionThrottle();
 
     /**
+     * Should player skulls with missing skin data be fetched from Mojang?
+     */
+    public boolean getFetchSkulls();
+
+    /**
      * Gets default ticks per animal spawns value.
      * <p>
      * <b>Example Usage:</b>
