From f59ab524726bf06d27ebf62b30f5f4ace6b40b7d Mon Sep 17 00:00:00 2001
From: cswhite2000 <18whitechristop@gmail.com>
Date: Mon, 12 Feb 2018 17:15:54 -0800
Subject: [PATCH] Add setting to disable shield knockback


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index 791ed020..c163cde4 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -367,6 +367,13 @@ public final class Bukkit {
         return server.getAdvancementsEnabled();
     }
 
+    /**
+     * @see Server#getDisableShieldKnockback()
+     */
+    public static boolean getDisableShieldKnockback() {
+        return server.getDisableShieldKnockback();
+    }
+
     /**
      * Gets default ticks per animal spawns value.
      * <p>
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 874a92c4..8a02d40a 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -293,6 +293,11 @@ public interface Server extends PluginMessageRecipient, BukkitRuntime, tc.oc.min
      */
     public boolean getAdvancementsEnabled();
 
+    /**
+     * Make shields less effective
+     */
+    public boolean getDisableShieldKnockback();
+
     /**
      * Gets default ticks per animal spawns value.
      * <p>
