From c90736f3af4999acfc0589cba3e61e6f50f1d207 Mon Sep 17 00:00:00 2001
From: Austin Mayes <austin@avicus.net>
Date: Sat, 1 Dec 2018 19:51:10 -0600
Subject: [PATCH] Allow block fall mechanics to be toggled


diff --git a/src/main/java/org/bukkit/Bukkit.java b/src/main/java/org/bukkit/Bukkit.java
index c163cde..0682348 100644
--- a/src/main/java/org/bukkit/Bukkit.java
+++ b/src/main/java/org/bukkit/Bukkit.java
@@ -361,6 +361,13 @@ public final class Bukkit {
     }
 
     /**
+     * @see Server#getContinueFalling()
+     */
+    public static boolean getContinueFalling() {
+        return server.getContinueFalling();
+    }
+
+    /**
      * @see Server#getAdvancementsEnabled()
      */
     public static boolean getAdvancementsEnabled() {
diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 456f53c..cda9c84 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -289,6 +289,11 @@ public interface Server extends PluginMessageRecipient, BukkitRuntime, tc.oc.min
     public boolean getLegacyKnockback();
 
     /**
+     * If, when shot through the air, falling blocks should continue falling when they land on a non-solid block
+     */
+    public boolean getContinueFalling();
+
+    /**
      * Should we announce advancements?
      */
     public boolean getAdvancementsEnabled();
