From aec25d01ca3df34c0d2b93a4c702285eaee54026 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sat, 24 Dec 2016 19:58:54 -0500
Subject: [PATCH] Chest lid API


diff --git a/src/main/java/org/bukkit/block/Chest.java b/src/main/java/org/bukkit/block/Chest.java
index 97dc781..e68ab8e 100644
--- a/src/main/java/org/bukkit/block/Chest.java
+++ b/src/main/java/org/bukkit/block/Chest.java
@@ -23,4 +23,10 @@ public interface Chest extends Container, Nameable {
      * @return the inventory
      */
     Inventory getBlockInventory();
+
+    boolean setLid(Lid lid);
+
+    Lid getLid();
+
+    enum Lid { OPEN, CLOSED, AUTO }
 }
