From 5da3494f2405854d5ff24360a08a38e818d5720d Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Thu, 29 Dec 2016 23:25:25 -0500
Subject: [PATCH] Don't tick empty worlds


diff --git a/src/main/java/org/bukkit/Server.java b/src/main/java/org/bukkit/Server.java
index 30e51979..2a09d829 100644
--- a/src/main/java/org/bukkit/Server.java
+++ b/src/main/java/org/bukkit/Server.java
@@ -193,6 +193,10 @@ public interface Server extends PluginMessageRecipient, BukkitRuntime, tc.oc.min
      */
     public boolean getGenerateStructures();
 
+    boolean getKeepSpawnInMemory();
+
+    boolean getTickEmptyWorlds();
+
     /**
      * Gets whether this server allows the End or not.
      *
