From bd5397203aa9e503101b32e6af958be563eda12d Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sun, 6 Mar 2016 03:43:13 -0500
Subject: [PATCH] Empty chunk API


diff --git a/src/main/java/org/bukkit/Chunk.java b/src/main/java/org/bukkit/Chunk.java
index 972e154e..4c721011 100644
--- a/src/main/java/org/bukkit/Chunk.java
+++ b/src/main/java/org/bukkit/Chunk.java
@@ -51,6 +51,11 @@ public interface Chunk extends Physical {
      */
     Set<Block> getBlocks(Material material);
 
+    /**
+     * Is this chunk entirely filled with air blocks?
+     */
+    boolean isEmpty();
+
     /**
      * Capture thread-safe read-only snapshot of chunk data
      *
