From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sat, 7 Jan 2017 04:14:27 -0500
Subject: [PATCH] Expand Physical API


diff --git a/src/main/java/org/bukkit/World.java b/src/main/java/org/bukkit/World.java
--- a/src/main/java/org/bukkit/World.java
+++ b/src/main/java/org/bukkit/World.java
@@ -529,6 +529,11 @@ public interface World extends PluginMessageRecipient, Metadatable, Physical {
      */
     public UUID getUID();
 
+    @Override
+    default UUID getWorldId() {
+        return getUID();
+    }
+
     /**
      * Gets the default spawn {@link Location} of this world
      *
