From d6100cce51bb2f25a046643fda04daf11370ba2d Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Mon, 1 Jun 2015 06:28:48 -0400
Subject: [PATCH] Improve BlockState API


diff --git a/src/main/java/org/bukkit/block/BlockState.java b/src/main/java/org/bukkit/block/BlockState.java
index 4b13d494..55aabf24 100644
--- a/src/main/java/org/bukkit/block/BlockState.java
+++ b/src/main/java/org/bukkit/block/BlockState.java
@@ -27,17 +27,49 @@ public interface BlockState extends Metadatable {
     Block getBlock();
 
     /**
+     * Get the material of this block state.
+     */
+    Material getMaterial();
+
+    /**
+     * Get the material and metadata of this block state.
+     */
+    MaterialData getMaterialData();
+
+    /**
+     * Set the material of this block state. If the new material is different
+     * from the current material, any metadata for the new material will be set
+     * to default values.
+     *
+     * This changes only the state of this object, not any blocks.
+     * Call {@link #update} to apply changes to actual blocks.
+     */
+    void setMaterial(Material material);
+
+    /**
+     * Set the material and metadata of this block state.
+     *
+     * This changes only the state of this object, not any blocks.
+     * Call {@link #update} to apply changes to actual blocks.
+     */
+    void setMaterialData(MaterialData materialData);
+
+    /**
      * Gets the metadata for this block state.
      *
      * @return block specific metadata
+     * @deprecated Confusing name
      */
+    @Deprecated
     MaterialData getData();
 
     /**
      * Gets the type of this block state.
      *
      * @return block type
+     * @deprecated Confusing name
      */
+    @Deprecated
     Material getType();
 
     /**
@@ -117,17 +149,22 @@ public interface BlockState extends Metadatable {
     Chunk getChunk();
 
     /**
-     * Sets the metadata for this block state.
+     * Sets the metadata ONLY for this block (NOT the material).
+     * The {@link Material} of the given data must match that of the block.
      *
      * @param data New block specific metadata
+     * @deprecated Confusing name, strange behavior
      */
+    @Deprecated
     void setData(MaterialData data);
 
     /**
      * Sets the type of this block state.
      *
      * @param type Material to change this block state to
+     * @deprecated Confusing name
      */
+    @Deprecated
     void setType(Material type);
 
     /**
