From 9fce7e184b0fbf7fead0503720fcb1a767525965 Mon Sep 17 00:00:00 2001
From: mrapple <tony@oc.tc>
Date: Sat, 17 Nov 2012 12:42:05 -0600
Subject: [PATCH] Make AsyncPlayerPreLoginEvent more versatile


diff --git a/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java b/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java
index 1d571889..78c5ea2a 100644
--- a/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java
+++ b/src/main/java/org/bukkit/event/player/AsyncPlayerPreLoginEvent.java
@@ -188,6 +188,10 @@ public class AsyncPlayerPreLoginEvent extends Event {
          * white list
          */
         KICK_WHITELIST,
+        /**
+         * The player's username was not verified with Mojang
+         */
+        KICK_VERIFY,
         /**
          * The player is not allowed to log in, for reasons undefined
          */
diff --git a/src/main/java/org/bukkit/event/player/PlayerPreLoginEvent.java b/src/main/java/org/bukkit/event/player/PlayerPreLoginEvent.java
index e8553f0f..aa49dfa2 100644
--- a/src/main/java/org/bukkit/event/player/PlayerPreLoginEvent.java
+++ b/src/main/java/org/bukkit/event/player/PlayerPreLoginEvent.java
@@ -151,6 +151,10 @@ public class PlayerPreLoginEvent extends Event {
          * white list
          */
         KICK_WHITELIST,
+        /**
+         * The player's username was not verified with Mojang
+         */
+        KICK_VERIFY,
         /**
          * The player is not allowed to log in, for reasons undefined
          */
