From 6e4e81ec6758a3eef6ddd3bd3595c098ef3eb614 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Mon, 21 Mar 2016 01:41:04 -0400
Subject: [PATCH] Multi-passenger API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index cdd832c7..309491d1 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -295,6 +295,23 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Locatable
     public boolean eject();
 
     /**
+     * @return true if the given entity is a passenger of this vehicle
+     */
+    boolean hasPassenger(Entity passenger);
+
+    /**
+     * Make the given entities passengers of this vehicle. Any existing
+     * passengers who are not in the list are ejected.
+     * @return given entities that could NOT become passengers for whatever reason
+     */
+    List<Entity> setPassengers(List<Entity> passengers);
+
+    /**
+     * Eject all passengers from this vehicle
+     */
+    void ejectAll();
+
+    /**
      * Returns the distance this entity has fallen
      *
      * @return The distance.
@@ -374,6 +391,12 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Locatable
     public boolean isInsideVehicle();
 
     /**
+     * Enter the given vehicle
+     * @return true if successful
+     */
+    boolean enterVehicle(Entity vehicle);
+
+    /**
      * Leave the current vehicle. If the entity is currently in a vehicle (and
      * is removed from it), true will be returned, otherwise false will be
      * returned.
