From ff75071b69944c057184840440f84385f6e05b85 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Sat, 3 Dec 2016 23:45:45 -0500
Subject: [PATCH] Pause entity API


diff --git a/src/main/java/org/bukkit/entity/Entity.java b/src/main/java/org/bukkit/entity/Entity.java
index 6264292..3b67098 100644
--- a/src/main/java/org/bukkit/entity/Entity.java
+++ b/src/main/java/org/bukkit/entity/Entity.java
@@ -559,4 +559,21 @@ public interface Entity extends Metadatable, CommandSender, Nameable, Locatable
      * @return reaction
      */
     PistonMoveReaction getPistonMoveReaction();
+
+    /**
+     * Pause/unpause this entity.
+     *
+     * Pausing an entity prevents various dynamic properties from advancing.
+     * The set of affected properties may expand in the future.
+     *
+     * Currently, only potion effect durations are affected.
+     */
+    void setPaused(boolean paused);
+
+    /**
+     * Is this entity paused?
+     *
+     * @see #setPaused
+     */
+    boolean isPaused();
 }
