From 63d5ed7554b23c34db7a0a262319d4f560678200 Mon Sep 17 00:00:00 2001
From: Jedediah Smith <jedediah@silencegreys.com>
Date: Thu, 6 Aug 2015 22:18:43 -0400
Subject: [PATCH] Skull skin API


diff --git a/src/main/java/org/bukkit/block/Skull.java b/src/main/java/org/bukkit/block/Skull.java
index f69f9477..3499a8f6 100644
--- a/src/main/java/org/bukkit/block/Skull.java
+++ b/src/main/java/org/bukkit/block/Skull.java
@@ -1,8 +1,11 @@
 package org.bukkit.block;
 
 import org.bukkit.OfflinePlayer;
+import org.bukkit.Skin;
 import org.bukkit.SkullType;
 
+import java.util.UUID;
+
 /**
  * Represents a captured state of a skull block.
  */
@@ -54,6 +57,21 @@ public interface Skull extends BlockState {
     public void setOwningPlayer(OfflinePlayer player);
 
     /**
+     * Set the owner and appearance of this skull. A skull with this data set
+     * does not need to fetch anything remotely.
+     *
+     * @param name Username of the skull's owner, can be null (appears in item tooltip)
+     * @param uuid UUID of the skull's owner
+     * @param skin Skull owner's skin
+     */
+    void setOwner(String name, UUID uuid, Skin skin);
+
+    /**
+     * Clear any owner data in this skull
+     */
+    void clearOwner();
+
+    /**
      * Gets the rotation of the skull in the world
      *
      * @return the rotation of the skull
diff --git a/src/main/java/org/bukkit/inventory/meta/SkullMeta.java b/src/main/java/org/bukkit/inventory/meta/SkullMeta.java
index c60860e1..661bcd39 100644
--- a/src/main/java/org/bukkit/inventory/meta/SkullMeta.java
+++ b/src/main/java/org/bukkit/inventory/meta/SkullMeta.java
@@ -1,7 +1,10 @@
 package org.bukkit.inventory.meta;
 
+import java.util.UUID;
+
 import org.bukkit.Material;
 import org.bukkit.OfflinePlayer;
+import org.bukkit.Skin;
 
 /**
  * Represents a skull ({@link Material#SKULL_ITEM}) that can have an owner.
@@ -55,5 +58,20 @@ public interface SkullMeta extends ItemMeta {
      */
     boolean setOwningPlayer(OfflinePlayer owner);
 
+    /**
+     * Set the owner and appearance of this skull. A skull with this data set
+     * does not need to fetch anything remotely.
+     *
+     * @param name Username of the skull's owner, can be null (appears in item tooltip)
+     * @param uuid UUID of the skull's owner
+     * @param skin Skull owner's skin
+     */
+    void setOwner(String name, UUID uuid, Skin skin);
+
+    /**
+     * Clear any owner data in this skull
+     */
+    void clearOwner();
+
     SkullMeta clone();
 }
